<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/v0.10.0/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.0" id="rtassel-0100">rTASSEL 0.10.0<a class="anchor" aria-label="anchor" href="#rtassel-0100"></a></h2>
<ul><li>Updated formula parsing:
<ul><li>Drop and keep traits by using <code>-</code> and <code>+</code>, respectively</li>
<li>Added keywords to drop/keep all covariate or factor traits:
<ul><li><code>I(cov)</code></li>
<li><code>I(fct)</code></li>
</ul></li>
</ul></li>
<li>Added <code><a href="../reference/stepwiseModelFitter.html">stepwiseModelFitter()</a></code> function:
<ul><li>Runs stepwise regression via TASSEL 5’s “Stepwise” plugin</li>
<li>Returns <code>AssociationResults</code> object</li>
</ul></li>
<li>Added new function <code><a href="../reference/readNumericGenotypeFromRMatrix.html">readNumericGenotypeFromRMatrix()</a></code>:
<ul><li>Converts formatted R matrix to TASSEL 5 numeric genotype</li>
</ul></li>
<li>Added experimental function <code><a href="../reference/readGenotype.html">readGenotype()</a></code>:
<ul><li>Reads data into new <code>TasselGenotype</code> class (future update)</li>
<li>Dynamically read genotype data based data type</li>
</ul></li>
<li>Added experimental function <code><a href="../reference/readPhenotype.html">readPhenotype()</a></code>:
<ul><li>Reads data into new <code>TasselPhenotype</code> class (future update)</li>
<li>Dynamically read phenotype data based data type</li>
</ul></li>
<li>Added deprecation warnings to the following methods:
<ul><li><code><a href="../reference/getPhenotypeDF.html">getPhenotypeDF()</a></code></li>
<li><code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code></li>
<li><code><a href="../reference/readGenotypeTableFromPath.html">readGenotypeTableFromPath()</a></code></li>
<li><code><a href="../reference/readGenotypeTableFromGigwa.html">readGenotypeTableFromGigwa()</a></code></li>
<li><code><a href="../reference/readGenotypePhenotype.html">readGenotypePhenotype()</a></code></li>
<li><code><a href="../reference/readPhenotypeFromDataFrame.html">readPhenotypeFromDataFrame()</a></code></li>
<li><code><a href="../reference/readPhenotypeFromPath.html">readPhenotypeFromPath()</a></code></li>
<li><code><a href="../reference/ldJavaApp.html">ldJavaApp()</a></code></li>
<li><code><a href="../reference/treeJavaApp.html">treeJavaApp()</a></code></li>
<li>
<strong>NOTE</strong>: these will be removed in future updates</li>
</ul></li>
<li>Updated vignettes:
<ul><li>Visualization section updates</li>
<li>Formula parsing</li>
<li>
<code>AssociationResults</code> and <code>PCAResults</code> object handling</li>
<li>Memory allocation guide</li>
<li>Numeric genotype handling</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.33" id="rtassel-0933">rTASSEL 0.9.33<a class="anchor" aria-label="anchor" href="#rtassel-0933"></a></h2>
<ul><li>Fixed typo in <code><a href="../reference/plotPCA.html">plotPCA()</a></code> error message</li>
<li>Add new function <code><a href="../reference/filterGenotypeTableBySiteName.html">filterGenotypeTableBySiteName()</a></code>:
<ul><li>Filters genotype tables using literal marker names/IDs</li>
</ul></li>
<li>Add new function <code><a href="../reference/mergeGenotypeTables.html">mergeGenotypeTables()</a></code>:
<ul><li>Merges multiple genotype tables by site values</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.32" id="rtassel-0932">rTASSEL 0.9.32<a class="anchor" aria-label="anchor" href="#rtassel-0932"></a></h2>
<ul><li>Updated <code><a href="../reference/tableReport.html">tableReport()</a></code> method dispatch for all <code>AssociationResults</code> objects:
<ul><li>Will now return default statistics output for all association results when running <code>tableReport(assocObj)</code> where <code>assocObj</code> is an object of type <code>AssociationResults</code>
</li>
</ul></li>
<li>Removed HDF5 file export support</li>
<li>Improved logic support for <code><a href="../reference/plotPCA.html">plotPCA()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.31" id="rtassel-0931">rTASSEL 0.9.31<a class="anchor" aria-label="anchor" href="#rtassel-0931"></a></h2>
<ul><li>Added new <code>PCAResults</code> class
<ul><li>Allows for more controlled access of data and simplified downstream functions for end users</li>
</ul></li>
<li>Add new function <code><a href="../reference/plotScree.html">plotScree()</a></code>:
<ul><li>Generates quick scree plots from <code>PCAResults</code> objects</li>
</ul></li>
<li>Add new function <code><a href="../reference/plotPCA.html">plotPCA()</a></code>:
<ul><li>Generates quick PCA plots from <code>PCAResults</code> objects</li>
<li>Allows for grouping from generated hierarchical clustering or grouping from metadata via the <code>metadata</code> parameter and subsequent <code>mCol</code> parameters.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.30" id="rtassel-0930">rTASSEL 0.9.30<a class="anchor" aria-label="anchor" href="#rtassel-0930"></a></h2>
<ul><li>Added new <code>AssociationResults</code> class
<ul><li>Allows for more controlled access of data and simplified downstream functions for end users</li>
</ul></li>
<li>Added new function <code><a href="../reference/plotManhattan.html">plotManhattan()</a></code>:
<ul><li>Supercedes older Manhattan plotting methods to work with new <code>AssociationResults</code> class.</li>
</ul></li>
<li>Added new function <code><a href="../reference/plotQQ.html">plotQQ()</a></code>:
<ul><li>Plotting function for QQ results from <code>AssociationResults</code> class</li>
</ul></li>
<li>Added new function <code><a href="../reference/plotManhattanQC.html">plotManhattanQC()</a></code>:
<ul><li>Plotting function and QC method for zoomed in regions of interest across genome</li>
</ul></li>
<li>Prior 3 functions also include interactive component that wraps <code>ggplot2</code> objects with PlotlyJS components</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.29" id="rtassel-0929">rTASSEL 0.9.29<a class="anchor" aria-label="anchor" href="#rtassel-0929"></a></h2>
<ul><li>Added genotype table summary methods:
<ul><li><code><a href="../reference/positionList.html">positionList()</a></code></li>
<li><code><a href="../reference/taxaSummary.html">taxaSummary()</a></code></li>
<li><code><a href="../reference/siteSummary.html">siteSummary()</a></code></li>
</ul></li>
<li>
<code>TasselGenotypePhenotype</code> objects containing genotype table data can now be coerced into R <code>matrix</code> objects using the function <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>
<ul><li>This will return a taxa x site matrix where taxa is the number of rows and sites is the number of columns.</li>
</ul></li>
<li>Added generalized join methods:
<ul><li><code><a href="../reference/intersectJoin.html">intersectJoin()</a></code></li>
<li><code><a href="../reference/unionJoin.html">unionJoin()</a></code></li>
<li>Joins phenotype data data based on taxa ID - similar to the TASSEL API</li>
</ul></li>
<li>Added read method for importing GIGWA data through <code>QBMS</code>:
<ul><li><code><a href="../reference/readGenotypeTableFromGigwa.html">readGenotypeTableFromGigwa()</a></code></li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.28" id="rtassel-0928">rTASSEL 0.9.28<a class="anchor" aria-label="anchor" href="#rtassel-0928"></a></h2>
<ul><li>Fixed <code>log4j</code> warning issue
<ul><li>This removes <code>log4j</code> warning messages when the <code><a href="../reference/startLogger.html">startLogger()</a></code> function is called.</li>
</ul></li>
<li>Removed <code>useRef</code> parameter from <code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code> function.
<ul><li>This is now automatically detected from the file input.</li>
<li>This fixes ref/alt allele vs major/minor allele encoding issues.</li>
</ul></li>
<li>Added Journal of Open Source Software citation for the <code>rTASSEL</code> package.
<ul><li>For citation information, use <code>utils::citation("rTASSEL")</code>
</li>
</ul></li>
<li>Added data object, <code>rtPaths</code>
<ul><li>Includes paths to external toy data for <code>rTASSEL</code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.27" id="rtassel-0927">rTASSEL 0.9.27<a class="anchor" aria-label="anchor" href="#rtassel-0927"></a></h2>
<ul><li>No significant updates in this version. This version is virtually identical to <code>0.9.26</code> and is for linking to Zenodo for archival purposes.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.26" id="rtassel-0926">rTASSEL 0.9.26<a class="anchor" aria-label="anchor" href="#rtassel-0926"></a></h2>
<ul><li>Bug fixes:
<ul><li>Fixed <code>r2</code> parameter bug in <code><a href="../reference/ldPlot.html">ldPlot()</a></code>
</li>
<li>Fixed space bugs in certain column names of data frame objects. <code>_</code> values now replace spaces.</li>
<li>Fixed <code>show()</code> method for <code>TasselDistanceMatrix</code> objects.</li>
</ul></li>
<li>Add new function:
<ul><li><code><a href="../reference/seqDiversity.html">seqDiversity()</a></code></li>
<li>Calculates diversity basic diversity metrics on genetic data.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.25" id="rtassel-0925">rTASSEL 0.9.25<a class="anchor" aria-label="anchor" href="#rtassel-0925"></a></h2>
<ul><li>Bug fixes:
<ul><li>Fixed character conversion bug in <code>DataFrame</code> object returns.</li>
</ul></li>
<li>
<code><a href="../reference/pca.html">pca()</a></code> can optionally report eigenvalues and eigenvectors as a list object.</li>
<li>Added new function:
<ul><li><code><a href="../reference/imputeNumeric.html">imputeNumeric()</a></code></li>
<li>Allows for numeric imputation of <code>GenotypeTable</code> objects.</li>
</ul></li>
<li>Added new function:
<ul><li><code>imputLDKNNi()</code></li>
<li>Allows for LD KNNi imputation of <code>GenotypeTable</code> objects.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.24" id="rtassel-0924">rTASSEL 0.9.24<a class="anchor" aria-label="anchor" href="#rtassel-0924"></a></h2>
<ul><li>Added new function:
<ul><li><code><a href="../reference/pca.html">pca()</a></code></li>
<li>Allows for user to run PCA on rTASSEL objects containing a <code>GenotypeTable</code> object.</li>
</ul></li>
<li>Added new function:
<ul><li><code><a href="../reference/mds.html">mds()</a></code></li>
<li>Allows for user to run MDS on <code>TasselDistanceMatrix</code> objects.</li>
</ul></li>
<li>Enhancements:
<ul><li>New summary print output for <code>TasselDistanceMatrix</code> objects.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.23" id="rtassel-0923">rTASSEL 0.9.23<a class="anchor" aria-label="anchor" href="#rtassel-0923"></a></h2>
<ul><li>Added new <code>TasselDistanceMatrix</code> class
<ul><li>Specified function (<code><a href="../reference/kinshipMatrix.html">kinshipMatrix()</a></code> and <code><a href="../reference/distanceMatrix.html">distanceMatrix()</a></code>) now return an object of type <code>TasselDistanceMatrix</code>.</li>
<li>Prevents console overload and freezing as seen with large distance matrix objects.</li>
<li>Now shows summary overview of matrix instead of Java object reference.</li>
<li>Generic functions <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code>, <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code>, <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol()</a></code>, and <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code> will return relative information similar to how these operate with <code>matrix</code> type objects.</li>
<li>Primitive function <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> now supersedes deprecated functions <code>kinshipToRMatrix()</code> and <code>distanceToRMatrix()</code>.</li>
<li>Prior functions that take in a kinship object will now take in this new class.</li>
</ul></li>
<li>Added new function:
<ul><li><code><a href="../reference/readTasselDistanceMatrix.html">readTasselDistanceMatrix()</a></code></li>
<li>Allows for user to read in delimited distance matrix stored in a flat file.</li>
</ul></li>
<li>Added new function:
<ul><li><code><a href="../reference/asTasselDistanceMatrix.html">asTasselDistanceMatrix()</a></code></li>
<li>Coerces a pairwise matrix (e.g. m x m dimensions) with the same column and row names of type <code>matrix</code> to an object of type <code>TasselDistanceMatrix</code>.</li>
</ul></li>
<li>Added new function:
<ul><li><code><a href="../reference/createTree.html">createTree()</a></code></li>
<li>interface to TASSEL’s tree creation methods</li>
<li>Allows for <code>Neighbor_Joining</code> and <code>UPGMA</code> methods</li>
</ul></li>
<li>Added new function:
<ul><li><code><a href="../reference/treeJavaApp.html">treeJavaApp()</a></code></li>
<li>wrapper for TASSEL’s interface to the Archaeopteryx Java tree Viewer</li>
<li>Implements same methods for tree creation as <code><a href="../reference/createTree.html">createTree()</a></code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.22" id="rtassel-0922">rTASSEL 0.9.22<a class="anchor" aria-label="anchor" href="#rtassel-0922"></a></h2>
<ul><li>Fix <code>manhattanPlot()</code> aesthetics:
<ul><li>Remove redundant marker labels from x-axis</li>
<li>Change x-axis label to <code>SNP Positions</code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.21" id="rtassel-0921">rTASSEL 0.9.21<a class="anchor" aria-label="anchor" href="#rtassel-0921"></a></h2>
<ul><li>Added new function:
<ul><li><code><a href="../reference/exportGenotypeTable.html">exportGenotypeTable()</a></code></li>
</ul></li>
<li>Added new vignette:
<ul><li>“Filtering Genotype Tables”</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.20" id="rtassel-0920">rTASSEL 0.9.20<a class="anchor" aria-label="anchor" href="#rtassel-0920"></a></h2>
<ul><li>Added new parameter to <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>
<ul><li>
<code>gRangesObj</code>: Filter genotype tables by using a <code>GRanges</code> object.</li>
</ul></li>
<li>Added new parameter to <code><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa()</a></code>
<ul><li>
<code>taxa</code>: Pass a vector of taxa IDs to filter genotype table by.</li>
</ul></li>
<li>Fixed <code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code> bug:
<ul><li>dosage array now returns <code>NA</code>s instead of <code>128</code> values.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.19" id="rtassel-0919">rTASSEL 0.9.19<a class="anchor" aria-label="anchor" href="#rtassel-0919"></a></h2>
<ul><li>Added two new parameters to <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>
<ul><li>
<code>removeMinorSNPStates</code>: Boolean; removes minor SNP states.</li>
<li>
<code>removeSitesWithIndels</code>: Boolean; removes sites with indels.</li>
</ul></li>
<li>Added better descriptive error handling for <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>
</li>
<li>Fixed <code>siteRangeFilterType</code> parameter bug in <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>. Now defaults to <code>none</code> when user does not specify filter type.</li>
<li>Added two new parameters to <code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code>
<ul><li>
<code>coerceDosageToInt</code>: Returns raw byte dosage array instead of integer from Java.</li>
<li>
<code>verbose</code>: Display console messages for large “memory-intensive” datasets.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.18" id="rtassel-0918">rTASSEL 0.9.18<a class="anchor" aria-label="anchor" href="#rtassel-0918"></a></h2>
<ul><li>Added functions to calculate linkage disequilibrium (LD)</li>
<li>Proposed LD functions:
<ul><li>
<code><a href="../reference/linkageDiseq.html">linkageDiseq()</a></code> - Returns TASSEL LD table report as data frame</li>
<li>
<code><a href="../reference/ldPlot.html">ldPlot()</a></code> - Returns static <code>ggplot2</code> plot</li>
<li>
<code><a href="../reference/ldJavaApp.html">ldJavaApp()</a></code> - Initiates TASSEL’s interactive LD viewer</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.17" id="rtassel-0917">rTASSEL 0.9.17<a class="anchor" aria-label="anchor" href="#rtassel-0917"></a></h2>
<ul><li>Added new function:
<ul><li><code>manhattanPlot()</code></li>
</ul></li>
<li>Removed tidyverse dependencies</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.16" id="rtassel-0916">rTASSEL 0.9.16<a class="anchor" aria-label="anchor" href="#rtassel-0916"></a></h2>
<ul><li>Added write to file parameters for:
<ul><li>assocModelFitter()</li>
</ul></li>
<li>Added p-value threshold parameters for:
<ul><li>assocModelFitter()</li>
</ul></li>
<li>Added thread usage paramters for:
<ul><li>assocModelFitter()</li>
</ul></li>
<li>Optimized table report to data frame generation</li>
<li>Added new filtration features for genotype tables via filterGenotypeTableSites()
<ul><li>parameters for variant sites</li>
<li>parameters for physical positions</li>
<li>filtration via chromsomome position files</li>
<li>filtration via BED file formats</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.13" id="rtassel-0913">rTASSEL 0.9.13<a class="anchor" aria-label="anchor" href="#rtassel-0913"></a></h2>
<ul><li>Added error checks for catching C stack usage errors for the following functions:
<ul><li>filterGenotypeTableSites()</li>
<li>filterGenotypeTableTaxa()</li>
</ul></li>
<li>Added NEWS file for tracking version updates.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.12" id="rtassel-0912">rTASSEL 0.9.12<a class="anchor" aria-label="anchor" href="#rtassel-0912"></a></h2>
<ul><li>Added new functions:
<ul><li><code>leaveOneFamilyOut()</code></li>
<li><code>genomicPredction()</code></li>
</ul></li>
<li>Fixed a bug where tibbles when passed through <code><a href="../reference/readPhenotypeFromDataFrame.html">readPhenotypeFromDataFrame()</a></code>, would cause errors.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

