<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • rTASSEL</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rTASSEL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.26</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/genotype_filtration.html">Filtering Genotype Tables</a>
    </li>
    <li>
      <a href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a>
    </li>
    <li>
      <a href="../articles/rtassel_walkthrough.html">Getting Started with rTASSEL</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.26" id="rtassel-0-9-26">rTASSEL 0.9.26<a class="anchor" aria-label="anchor" href="#rtassel-0-9-26"></a></h2>
<ul><li>Bug fixes:<ul><li>Fixed <code>r2</code> parameter bug in <code><a href="../reference/ldPlot.html">ldPlot()</a></code>
</li>
<li>Fixed space bugs in certain column names of data frame objects. <code>_</code> values now replace spaces.</li>
<li>Fixed <code>show()</code> method for <code>TasselDistanceMatrix</code> objects.</li>
</ul></li>
<li>Add new function:<ul><li><code><a href="../reference/seqDiversity.html">seqDiversity()</a></code></li>
<li>Calculates diversity basic diversity metrics on genetic data.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.25" id="rtassel-0-9-25">rTASSEL 0.9.25<a class="anchor" aria-label="anchor" href="#rtassel-0-9-25"></a></h2>
<ul><li>Bug fixes:<ul><li>Fixed character conversion bug in <code>DataFrame</code> object returns.</li></ul></li>
<li>
<code><a href="../reference/pca.html">pca()</a></code> can optionally report eigenvalues and eigenvectors as a list object.</li>
<li>Added new function:<ul><li><code><a href="../reference/imputeNumeric.html">imputeNumeric()</a></code></li>
<li>Allows for numeric imputation of <code>GenotypeTable</code> objects.</li>
</ul></li>
<li>Added new function:<ul><li><code>imputLDKNNi()</code></li>
<li>Allows for LD KNNi imputation of <code>GenotypeTable</code> objects.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.24" id="rtassel-0-9-24">rTASSEL 0.9.24<a class="anchor" aria-label="anchor" href="#rtassel-0-9-24"></a></h2>
<ul><li>Added new function:<ul><li><code><a href="../reference/pca.html">pca()</a></code></li>
<li>Allows for user to run PCA on rTASSEL objects containing a <code>GenotypeTable</code> object.</li>
</ul></li>
<li>Added new function:<ul><li><code><a href="../reference/mds.html">mds()</a></code></li>
<li>Allows for user to run MDS on <code>TasselDistanceMatrix</code> objects.</li>
</ul></li>
<li>Enhancements:<ul><li>New summary print output for <code>TasselDistanceMatrix</code> objects.</li></ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.23" id="rtassel-0-9-23">rTASSEL 0.9.23<a class="anchor" aria-label="anchor" href="#rtassel-0-9-23"></a></h2>
<ul><li>Added new <code>TasselDistanceMatrix</code> class<ul><li>Specified function (<code><a href="../reference/kinshipMatrix.html">kinshipMatrix()</a></code> and <code><a href="../reference/distanceMatrix.html">distanceMatrix()</a></code>) now return an object of type <code>TasselDistanceMatrix</code>.</li>
<li>Prevents console overload and freezing as seen with large distance matrix objects.</li>
<li>Now shows summary overview of matrix instead of Java object reference.</li>
<li>Generic functions <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code>, <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code>, <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol()</a></code>, and <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code> will return relative information similar to how these operate with <code>matrix</code> type objects.</li>
<li>Primitive function <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> now supersedes deprecated functions <code>kinshipToRMatrix()</code> and <code>distanceToRMatrix()</code>.</li>
<li>Prior functions that take in a kinship object will now take in this new class.</li>
</ul></li>
<li>Added new function:<ul><li><code><a href="../reference/readTasselDistanceMatrix.html">readTasselDistanceMatrix()</a></code></li>
<li>Allows for user to read in delimited distance matrix stored in a flat file.</li>
</ul></li>
<li>Added new function:<ul><li><code><a href="../reference/asTasselDistanceMatrix.html">asTasselDistanceMatrix()</a></code></li>
<li>Coerces a pairwise matrix (e.g. m x m dimensions) with the same column and row names of type <code>matrix</code> to an object of type <code>TasselDistanceMatrix</code>.</li>
</ul></li>
<li>Added new function:<ul><li><code><a href="../reference/createTree.html">createTree()</a></code></li>
<li>interface to TASSEL’s tree creation methods</li>
<li>Allows for <code>Neighbor_Joining</code> and <code>UPGMA</code> methods</li>
</ul></li>
<li>Added new function:<ul><li><code><a href="../reference/treeJavaApp.html">treeJavaApp()</a></code></li>
<li>wrapper for TASSEL’s interface to the Archaeopteryx Java tree Viewer</li>
<li>Implements same methods for tree creation as <code><a href="../reference/createTree.html">createTree()</a></code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.22" id="rtassel-0-9-22">rTASSEL 0.9.22<a class="anchor" aria-label="anchor" href="#rtassel-0-9-22"></a></h2>
<ul><li>Fix <code><a href="../reference/manhattanPlot.html">manhattanPlot()</a></code> aesthetics:<ul><li>Remove redundant marker labels from x-axis</li>
<li>Change x-axis label to <code>SNP Positions</code>
</li>
</ul></li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.21" id="rtassel-0-9-21">rTASSEL 0.9.21<a class="anchor" aria-label="anchor" href="#rtassel-0-9-21"></a></h2>
<ul><li>Added new function:<ul><li><code><a href="../reference/exportGenotypeTable.html">exportGenotypeTable()</a></code></li></ul></li>
<li>Added new vignette:<ul><li>“Filtering Genotype Tables”</li></ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.20" id="rtassel-0-9-20">rTASSEL 0.9.20<a class="anchor" aria-label="anchor" href="#rtassel-0-9-20"></a></h2>
<ul><li>Added new parameter to <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code><ul><li>
<code>gRangesObj</code>: Filter genotype tables by using a <code>GRanges</code> object.</li></ul></li>
<li>Added new parameter to <code><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa()</a></code><ul><li>
<code>taxa</code>: Pass a vector of taxa IDs to filter genotype table by.</li></ul></li>
<li>Fixed <code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code> bug:<ul><li>dosage array now returns <code>NA</code>s instead of <code>128</code> values.</li></ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.19" id="rtassel-0-9-19">rTASSEL 0.9.19<a class="anchor" aria-label="anchor" href="#rtassel-0-9-19"></a></h2>
<ul><li>Added two new parameters to <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code><ul><li>
<code>removeMinorSNPStates</code>: Boolean; removes minor SNP states.</li>
<li>
<code>removeSitesWithIndels</code>: Boolean; removes sites with indels.</li>
</ul></li>
<li>Added better descriptive error handling for <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>
</li>
<li>Fixed <code>siteRangeFilterType</code> parameter bug in <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>. Now defaults to <code>none</code> when user does not specify filter type.</li>
<li>Added two new parameters to <code><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code><ul><li>
<code>coerceDosageToInt</code>: Returns raw byte dosage array instead of integer from Java.</li>
<li>
<code>verbose</code>: Display console messages for large “memory-intensive” datasets.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.18" id="rtassel-0-9-18">rTASSEL 0.9.18<a class="anchor" aria-label="anchor" href="#rtassel-0-9-18"></a></h2>
<ul><li>Added functions to calculate linkage disequilibrium (LD)</li>
<li>Proposed LD functions:<ul><li>
<code><a href="../reference/linkageDiseq.html">linkageDiseq()</a></code> - Returns TASSEL LD table report as data frame</li>
<li>
<code><a href="../reference/ldPlot.html">ldPlot()</a></code> - Returns static <code>ggplot2</code> plot</li>
<li>
<code><a href="../reference/ldJavaApp.html">ldJavaApp()</a></code> - Initiates TASSEL’s interactive LD viewer</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.17" id="rtassel-0-9-17">rTASSEL 0.9.17<a class="anchor" aria-label="anchor" href="#rtassel-0-9-17"></a></h2>
<ul><li>Added new function:<ul><li><code><a href="../reference/manhattanPlot.html">manhattanPlot()</a></code></li></ul></li>
<li>Removed tidyverse dependencies</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.16" id="rtassel-0-9-16">rTASSEL 0.9.16<a class="anchor" aria-label="anchor" href="#rtassel-0-9-16"></a></h2>
<ul><li>Added write to file parameters for:<ul><li>assocModelFitter()</li></ul></li>
<li>Added p-value threshold parameters for:<ul><li>assocModelFitter()</li></ul></li>
<li>Added thread usage paramters for:<ul><li>assocModelFitter()</li></ul></li>
<li>Optimized table report to data frame generation</li>
<li>Added new filtration features for genotype tables via filterGenotypeTableSites()<ul><li>parameters for variant sites</li>
<li>parameters for physical positions</li>
<li>filtration via chromsomome position files</li>
<li>filtration via BED file formats</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.13" id="rtassel-0-9-13">rTASSEL 0.9.13<a class="anchor" aria-label="anchor" href="#rtassel-0-9-13"></a></h2>
<ul><li>Added error checks for catching C stack usage errors for the following functions:<ul><li>filterGenotypeTableSites()</li>
<li>filterGenotypeTableTaxa()</li>
</ul></li>
<li>Added NEWS file for tracking version updates.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.12" id="rtassel-0-9-12">rTASSEL 0.9.12<a class="anchor" aria-label="anchor" href="#rtassel-0-9-12"></a></h2>
<ul><li>Added new functions:<ul><li><code>leaveOneFamilyOut()</code></li>
<li><code>genomicPredction()</code></li>
</ul></li>
<li>Fixed a bug where tibbles when passed through <code><a href="../reference/readPhenotypeFromDataFrame.html">readPhenotypeFromDataFrame()</a></code>, would cause errors.</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brandon Monier, Terry Casstevens, Ed Buckler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

