---
title: "Getting Started with rTASSEL"
author: "Brandon Monier"
date:   "2021-03-10"
output:
    BiocStyle::html_document:
      toc: true
      fig_caption: true
      toc_float: true
      number_sections: false
vignette: >
    %\VignetteIndexEntry{Getting Start with rTASSEL}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
    %\usepackage[utf8]{inputenc}
    %\usepackage{float}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    fig.path='figure/graphics-',
    cache.path='cache/graphics-',
    fig.align='center',
    external=TRUE,
    echo=TRUE,
    warning=FALSE
    # fig.pos="H"
)
```



# Introduction

## Overview
Thanks for checking out rTASSEL! In this document, we will go over the
functionalities used to work with the TASSEL software via R.

TASSEL is a software package used to evaluate traits associations, evolutionary
patterns, and linkage disequilibrium. Strengths of this software include:

1. The opportunity for a number of new and powerful statistical approaches to
   association mapping such as a General Linear Model (GLM) and Mixed Linear
   Model (MLM). MLM is an implementation of the technique which our lab's
   published Nature Genetics paper - Unified Mixed-Model Method for
   Association Mapping - which reduces Type I error in association mapping
   with complex pedigrees, families, founding effects and population structure.

2. An ability to handle a wide range of indels (insertion & deletions). Most
   software ignore this type of polymorphism; however, in some species
   (like maize), this is the most common type of polymorphism.

More information can be found in the following paper:

Bradbury PJ, Zhang Z, Kroon DE, Casstevens TM, Ramdoss Y, Buckler ES. (2007)
*TASSEL: Software for association mapping of complex traits in diverse samples.* Bioinformatics 23:2633-2635.

Detailed documentation and source code can be found on our website:

https://www.maizegenetics.net/tassel

## Motivation
The main goal of developing this package is to construct an R-based front-end
to connect to a variety of highly used TASSEL methods and analytical tools.
By using R as a front-end, we aim to utilize a *unified* scripting workflow that
exploits the analytical prowess of TASSEL in conjunction with R's popular
data handling and parsing capabilities without ever having the user to switch
between these two environments.

## Disclaimer
Due to the *experimental* nature of this package's lifecycle, end
functionalities are prone to change after end-user input is obtained in the
near future.



# Installation and Preliminary Steps

## Prerequisites - installing rJava
Since TASSEL is written in Java, Java JDK will need to be installed on your
machine. Additionally, for R to communicate with Java, the R package `rJava`
will need to be installed. Detailed information can be found using the
following links, depending on your OS:

* [Linux](https://datawookie.netlify.com/blog/2018/02/installing-rjava-on-ubuntu/)
* [macOS](https://zhiyzuo.github.io/installation-rJava/)
* [Windows](https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/)


## Problems with rJava if you have upgraded Java
On a MAC: if you previously had `rJava` working through RStudio, then you
upgraded your Java and it now longer works, try the following:

At the command line type:

```
R CMD javareconf
```

Then check for a left over symbolic link via:

```
ls -ltr /usr/local/lib/libjvm.dylib
```

If the link exists, remove it, then create it fresh via these commands:

```
rm /usr/local/lib/libjvm.dylib
sudo ln -s $(/usr/libexec/java_home)/lib/server/libjvm.dylib /usr/local/lib
```

You should now be able to enter RStudio and setup rJava.


## Install from BitBucket
After you have `rJava` up and running on your machine, install `rTASSEL` by
installing the source code from our BitBucket repository:

```{r, eval=FALSE, echo=TRUE}
if (!require("devtools")) install.packages("devtools")
devtools::install_bitbucket(
    repo = "bucklerlab/rtassel",
    ref = "master",
    build_vignettes = TRUE
)
```

After source code has been compiled, the package can be loaded using:

```{r, eval=TRUE, echo=TRUE}
library(rTASSEL)
```

Or, if you want to use a function without violating your environment you can
use `rTASSEL::<function>`, where `<function>` is an `rTASSEL` function.


## Problems installing if you have both 32- and 64-bit architecture installed for R
**NOTE:** If you are using a machine that has both architectures installed for R, 
you might run into problems pulling code using `devtools`. If this is the case,
one solution would be to add the parameter `--no-multiarch` option in 
`INSTALL_opts`. This will force building the package for your currently running 
R version:

```{r, eval=FALSE, echo = TRUE}
devtools::install_bitbucket(
    repo = "bucklerlab/rTASSEL",
    ref = "master",
    build_vignettes = FALSE,
    INSTALL_opts = "--no-multiarch"
)
```


## Preliminary steps

### Setting Memory
Since genome-wide association analyses can use up a lot of computational
resources, memory allocation to `rTASSEL` can be modified. To change the amount
of memory, use the base `options()` function and modify the following parameter:

```{r, eval=FALSE, echo=TRUE}
options(java.parameters = c("-Xmx<memory>", "-Xms<memory>"))
```

Replace `<memory>` with a specified unit of memory. For example, if I want to
allocate a maximum of 6 GB of memory for my operations, I would use the input
`"-Xmx6g"`, where `g` stands for gigabyte (GB). More information about memory
allocation can be found [here](https://stackoverflow.com/questions/14763079/what-are-the-xms-and-xmx-parameters-when-starting-jvm).

**NOTE:** Setting Java memory options for `rTASSEL` and any `rJava`-related packages
*needs* to be set *before* loading the `rTASSEL` package!

### The importance of logging your progress
Before we begin analyzing data, optional parameters can be set up to make
`rTASSEL` more efficient. To prevent your R console from being overloaded with
TASSEL logging information, **it is highly recommended that you start a logging
file**. This file will house all of TASSEL's logging output which is
beneficial for debugging and tracking the progress of your analytical workflow.
To start a logging file, use the following command:

```{r, eval=FALSE, echo=TRUE}
rTASSEL::startLogger(fullPath = NULL, fileName = NULL)
```

If the `rTASSEL::startLogger()` parameters are set to `NULL`, the logging file
will be created in your current working directory. If you are unsure of what
your working directory is in R, use the base `getwd()` command.


Additionally, since this is a general walkthrough, certain intricaces of each
function may glossed over. If you would like to study a function in full,
refer to the R documentation by using `?<function>` in the console, where
`<function>` is an `rTASSEL`-based function.



# Reading Data

## Overview
Like TASSEL, `rTASSEL` will read two main types of data:

* Genotype data
* Phenotype data

This data can be read in several different ways. In the following examples,
we will demonstrate various ways genotype and phenotype information can be
loaded into `rTASSEL` objects.


## Loading genotype data

### From a path
Currently, reading in genotype data to `rTASSEL` is based off of file
locations as paths. Genotype/sequencing data can be stored in a variety of
formats. `rTASSEL` can read and store a wide variety of file types:

* hapmap (HMP)
* HDF5 (hierarchical data format version 5)
* VCF (variant call format)
* Plink

To load this genotype data, simply store your file location as a string object
in R. For this example, we will load two toy data sets - one being a VCF file
and the other being a hapmap file. These data sets can be accessed via the
`rTASSEL` package itself:

```{r, eval=TRUE, echo=TRUE}
# Load hapmap data
genoPathHMP <- system.file(
    "extdata",
    "mdp_genotype.hmp.txt",
    package = "rTASSEL"
)
genoPathHMP

# Load VCF data
genoPathVCF <- system.file(
    "extdata",
    "maize_chr9_10thin40000.recode.vcf",
    package = "rTASSEL"
)
genoPathVCF
```

Now that we have the file paths to this data, we can pass this to TASSEL and
create a formal `TasselGenotypePhenotype` class object in R using the
following:

```{r, eval=TRUE, echo=TRUE}
# Load in hapmap file
tasGenoHMP <- rTASSEL::readGenotypeTableFromPath(
    path = genoPathHMP
)

# Load in VCF file
tasGenoVCF <- rTASSEL::readGenotypeTableFromPath(
    path = genoPathVCF
)
```

When we call these objects, a summary of the data will be posted to the R
console:

```{r, eval=TRUE, echo=TRUE}
tasGenoHMP
```

This summary details the number of Taxa (`Taxa`) and marker positions
(`Positions`) within the data set. Additionally, since we can load both
genotype and phenotype information into this object, a helpful check will be
displayed to show what is populating the object (`[x] or [ ]`).

### Additional information about TasselPhenotypeGenotype data sets
In general, this S4 class data object houses "slot" information relating to
TASSEL/Java pointers of the respective data.

```{r, eval=TRUE, echo=TRUE}
class(tasGenoHMP)
slotNames(tasGenoHMP)
```

Technically, this object does not contain the full information of the data
represented in R space, but merely contains addresses to the memory store of
the reference TASSEL object ID. For example, if we wanted to extract the
`GenotypeTable` with the S4 `@` operator, we would get something that looks
like this:

```{r, eval=TRUE, echo=TRUE}
tasGenoHMP@jGenotypeTable
```

This entity is a `rJava` internal identifier. It isn't until we call
downstream `rTASSEL` functions where we will bring the TASSEL data into the R
environment.


## Loading phenotype data

### From a path
Similar to reading in genotype data, phenotype data can also be read in via
paths. If you already have preconstructed phenotype data in a file, this
option will most likely work best for you. One caveat to this is how the
data file is constructed in terms of columns and trait data for TASSEL
analyses. More information about how these files can be found at this
[link](https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/Load/Load)
under the *Numerical Data* section.

Loading this type of data is very similar to how genotype data is loaded.
here, we will use the `rTASSEL::readPhenotypeFromPath()` function:

```{r, eval=TRUE, echo=TRUE}
# Read from phenotype path
phenoPath  <- system.file("extdata", "mdp_traits.txt", package = "rTASSEL")
phenoPath

# Load into rTASSEL `TasselGenotypePhenotype` object
tasPheno <- rTASSEL::readPhenotypeFromPath(
    path = phenoPath
)

# Inspect object
tasPheno
```

The object output is very similar to the genotype table output with some minor
additions to which traits are displayed in the file.

### From an R data frame
In some cases you might want to first modify your phenotype data set in R and
then load it into the TASSEL environment. If you wish to choose this route, you
will need to use the `rTASSEL::readPhenotypeFromDataFrame()` function along
with a couple of parameters. First, we will construct an R data frame and load
it with this function:

```{r, eval=TRUE, echo=TRUE}
# Create phenotype data frame
phenoDF <- read.table(phenoPath, header = TRUE)
colnames(phenoDF)[1] <- "Taxon"

# Inspect first few rows
head(phenoDF)

# Load into rTASSEL `TasselGenotypePhenotype` object
tasPhenoDF <- rTASSEL::readPhenotypeFromDataFrame(
    phenotypeDF = phenoDF,
    taxaID = "Taxon",
    attributeTypes = NULL
)

# Inspect new object
tasPhenoDF
```

The `phenotypeDF` parameter is for the R data frame object. The `taxaID`
parameter is needed to determine which column of your data frame is your
TASSEL taxa data. The final parameter (`attributeTypes`) is optional. If this
parameter is set to `NULL`, all remaining data frame columns will be classified
as TASSEL `data` types. If this is not the case for your data (e.g. if you
have covariate or factor data in your experiment), **you will need to
specify which columns are what TASSEL data type** (i.e. `data`, `covariate`,
or `factor`). This will have to be passed as an R vector of string elements
(e.g. `c("data", "factor", "covariate")`). *Currently*, this data type needs to
be entered in the same order as they are found in the data frame.


## Loading genotype and phenotype data simultaneously
In association studies, we are interested in combining our genotype and
phenotype data. To usually run this operation in TASSEL, an intersect
combination between the two data sets is needed. To run this in `rTASSEL`, we
can use the `rTASSEL::readGenotypePhenotype()` function. The parameter input
needed for this function is, of course, a genotype and phenotype object. For
genotype input, the following can be used:

* a path to a genotype file
* a prior `TasselGenotypePhenotype` object

For phenotype input, the following can be used:

* a path to a phenotype data set,
* a prior `TasselGenotypePhenotype` object
* an R data frame

For example, if we wanted to read the prior `TasselGenotypePhenotype`
genotype and phenotype objects from earlier:

```{r, eval=TRUE, echo=TRUE}
tasGenoPheno <- rTASSEL::readGenotypePhenotype(
    genoPathOrObj = tasGenoHMP,
    phenoPathDFOrObj = tasPheno
)
tasGenoPheno
```

We can also use a combination of the above parameter options (e.g. load a
genotype path and a phenotype data frame, etc.). One caveat though, **if you
load in a phenotype data frame object with this function, the prior parameters
from the** `rTASSEL::readPhenotypeFromDataFrame` **will be needed** (i.e. the
`taxaID` and `attributeTypes` parameters):

```{r, eval=TRUE, echo=TRUE}
tasGenoPhenoDF <- rTASSEL::readGenotypePhenotype(
    genoPathOrObj = genoPathHMP,
    phenoPathDFOrObj = phenoDF,
    taxaID = "Taxon",
    attributeTypes = NULL
)
tasGenoPhenoDF
```


## Reading kinship data
rTASSEL also provides users the ability to read in delimited "flat-file" 
kinship objects as a `TasselDistanceMatrix` object using the function,
`readTasselDistanceMatrix()`:

```{r, eval=TRUE, echo=TRUE}
## Get toy kinship data from package ----
kinshipPath <- system.file(
  "extdata", 
  "mdp_kinship.txt", 
  package = "rTASSEL"
)

## Read ----
readTasselDistanceMatrix(kinshipPath)
```



# Extracting data from rTASSEL objects

## Get genotype data
If you want to bring in genotype data into the R environment, you can use
the `rTASSEL::getSumExpFromGenotypeTable()` function. All this function needs
is a `TasselGenotypePhenotype` class object *containing a genotype table*:

```{r, eval=TRUE, echo=TRUE}
tasSumExp <- rTASSEL::getSumExpFromGenotypeTable(
    tasObj = tasGenoPheno
)
tasSumExp
```

As you can see above, the gentoype object is returned as a
`SummarizedExperiment` class R object. More information about these objects can
be found [here](http://www.bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

From this object, we extract taxa information and their respective TASSEL
integer locations:

```{r, eval=TRUE, echo=TRUE}
SummarizedExperiment::colData(tasSumExp)
```

We can also extract the allelic marker data using
`SummarizedExperiment::rowData()`:

```{r, eval=TRUE, echo=TRUE}
SummarizedExperiment::rowData(tasSumExp)
```

And get marker coordinates using `SummarizedExperiment::rowRanges()`

```{r, eval=TRUE, echo=TRUE}
SummarizedExperiment::rowRanges(tasSumExp)
```


## Extract phenotype data
If you want to bring in phenotype data into the R environment, you can use
the `rTASSEL::getPhenotypeDF()` function. All this function needs
is a `TasselGenotypePhenotype` class object *containing a phenotype table*:

```{r, eval=TRUE, echo=TRUE}
tasExportPhenoDF <- rTASSEL::getPhenotypeDF(
    tasObj = tasGenoPheno
)
tasExportPhenoDF
```

As shown above, an R `tibble`-based data frame is exported with converted
data types translated from TASSEL. See the following table what TASSEL data
types are tranlated into within the R environment:

| TASSEL Data | Converted R Data type |
|:------------|:----------------------|
| taxa        | character             |
| data        | numeric               |
| covariate   | numeric               |
| factor      | factor                |



# Filtering genotype data

**NOTE**: This is just a "snapshot" of how we can filter genotype information
in rTASSEL. For more information, please see the additional vignette, 
*"Filtering Genotype Tables"*.

Prior to association analyses, filtration of genotype data may be necessary.
In TASSEL, this accomplished through the Filter menu using two primary plugins:

* Filter Site Builder plugin
* Filter Taxa Builder plugin

In `rTASSEL`, this can also be accomplished using the follwing functions:

* `rTASSEL::filterGenotypeTableSites()`
* `rTASSEL::filterGenotypeTableTaxa()`

These objects take a `TasselGenotypePhenotype` class object. For example, in
our genotype data set, if we want to remove monomorphic and low coverage sites,
we could use the following parameters in `rTASSEL::filterGenotypeTableSites()`:

```{r, eval=TRUE, echo=TRUE}
tasGenoPhenoFilt <- rTASSEL::filterGenotypeTableSites(
    tasObj = tasGenoPheno,
    siteMinCount = 150,
    siteMinAlleleFreq = 0.05,
    siteMaxAlleleFreq = 1.0,
    siteRangeFilterType = "none"
)
tasGenoPhenoFilt
```

We can then compare this to our original pre-filtered data set:

```{r, eval=TRUE, echo=TRUE}
tasGenoPheno
```

These functions can work on any `TasselGenotypePhenotype` class object that
contains genotypic data, regardless of single or combined TASSEL objects.



# Analysis - Relatedness

## Create a kinship matrix object
In TASSEL, for mixed linear model analyses, a kinship matrix calculated from
genotype data is necessary. This can be accomplished by calculating a kinship
TASSEL object using the function `kinshipMatrix()`. The main
parameter input is a `TasselGenotypePhenotype` class object that contains a
genotype data set:

```{r, eval=TRUE, echo=TRUE}
tasKin <- kinshipMatrix(tasObj = tasGenoPheno)
```

This function allows for several types of algorithm to used using the `method`
parameter. More info about these methods can be found
[here](https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/Kinship/Kinship).


## Calculate a distance matrix
Very similar to kinship matrix calculation, a distance matrix can also be
calculated using genotype data using the function `distanceMatrix()`:

```{r, eval=TRUE, echo=TRUE}
tasDist <- distanceMatrix(tasObj = tasGenoPheno)
```


## `TasselDistanceMatrix` objects

### Overview
The prior two functions will generate a pairwise matrix (e.g. $m \times m$ 
dimensions). The return object is an rTASSEL class, `TasselDistanceMatrix`.
When we inspect the prior object we will see something like this:

```{r, eval=TRUE, echo=TRUE}
tasKin
```

This will showcase the first four rows and columns and the last row and column
if the distance matrix exceeds 5 dimensions (*which it probably will*).

This object, similar to the `TasselGenotypePhenotype` class, essentially holds
pointers to the Java/TASSEL object in memory. Despite this, we can still use
some base R methods similar to how we handle `matrix` objects:

```{r, eval=TRUE, echo=TRUE}
library(magrittr)

tasKin %>% colnames() %>% head()
tasKin %>% rownames() %>% head()

tasKin %>% dim()

tasKin %>% nrow()
tasKin %>% ncol()
```

### Coercion
If we want to use additional R methods (e.g. plotting, new models, etc.), we
can coerce this object to a general R data object, in this case, a `matrix`
object using the base method `as.matrix()`:

```{r, eval=TRUE, echo=TRUE}
library(magrittr)

tasKinR <- tasKin %>% as.matrix()

## Inspect first 5 rows and columns ----
tasKinR[1:5, 1:5]
```

We can also coerce a pairwise `matrix` object to a `TasselDistanceMatrix` 
object using rTASSEL's function `asTasselDistanceMatrix()`:

```{r, eval=TRUE, echo=TRUE}
library(magrittr)

## Create a dummy pairwise matrix object ----
set.seed(123)
m <- 10
s <- matrix(rnorm(100), m)
s[lower.tri(s)] <- t(s)[lower.tri(s)]
diag(s) <- 2

## Add sample IDs ----
colnames(s) <- rownames(s) <- paste0("s_", seq_len(m))

testTasselDist <- s %>% asTasselDistanceMatrix()
testTasselDist
```



# Analysis - Association

## Overview
One of TASSEL's most powerful functionalities is its capability of performing
a variety of different association modeling techniques. If you have started
reading the walkthrough here it is *strongly suggested that you read the other
components of this walkthrough since the following parameters require what we
have previously created!*

If you are not familar with these methods, more information about how
these operate in base TASSEL can be found at following links:

* [BLUE/GLM](https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/GLM/GLM)
* [MLM](https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/MLM/MLM)
* [Fast Association](https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/FastAssociation/FastAssociation)

The `rTASSEL::assocModelFitter()` function has several primary components:

* `tasObj`: a `TasselGenotypePhenotype` class R object
* `formula`: an R-based linear model formula
* `fitMarkers`: a boolean parameter to differentiate between BLUE and GLM
  analyses
* `kinship`: a TASSEL kinship object
* `fastAssociation`: a boolean parameter for data sets that have many traits

Probably the most important concept of this function is `formula` parameter.
If you are familar with standard R linear model functions, this concept is
fairly similar. In TASSEL, a linear model is composed of the following scheme:

```
y ~ A
```

...where `y` is any TASSEL `data` type and `A` is any TASSEL `covariate`
and / or `factor` types:

```
<data> ~ <covariate> and/or <factor>
```

This model can be written out in several ways. With an example phenotype data,
we can have the following variables that are represented in TASSEL in the
following way:

* `Taxon` `<taxa>`
* `EarHT` `<data>`
* `dpoll` `<data>`
* `EarDia` `<data>`
* `location` `<factor>`
* `Q1` `<covariate>`
* `Q2` `<covariate>`
* `Q3` `<covariate>`

Using this data, we could write out the following formula in R

```
list(EarHT, dpoll, EarDia) ~ location + Q1 + Q2 + Q3
```

In the above example, we use a base `list()` function to indicate analysis
on multiple numeric data types. For covariate and factor information, we use
`+` operator. One problem with this implementation is that it can become
cumbersome and prone to error if we want to analyze the entirety of a large
data set or all data and/or factor and covariate types.

A work around for this problem is to utilize a special character to indicate
all elements within the model (`.`). By using the `.` operator we can simplify
the above model into the following:

```
. ~ .
```

This indicates we want to analyze *the whole* data set and leave nothing out.
If we want to analyze all data types and only a handful of factor and/or
covariates, we can use something like this:

```
. ~ location + Q1 + Q2
```

Or vice-versa:

```
list(EarHT, dpoll) ~ .
```

Take note we can be very specific with what we want to include in our trait
model! In the above example we have deliberately left out `EarDia` from our
model.

Additionally, we can also fit marker and kinship data to our model which can
change our analytical methods. Since these options in TASSEL are binary,
additional parameters are passed for this function. In this case,
genotype/marker data is fitted using the `fitMarker` parameter and kinship is
fitted using the `kinship` parameter.

Fast Association implements methods described by Shabalin (2012). This method
provides an ordinary least squares solution for fixed effect models. For this
method to proper work it is necessary that your have:

* *No missing data in your phenotype data set*
* *Phenotypes and genotypes have been merged using an intersect join*. Since
  this is currently the only option of join genotype and phenotype data, you
  do not have to worry about this *for now*.

**NOTE**: since we are working with "toy" data, empirical insight will not
be elucidated upon in the following steps. This is simply to show the user how
properly use these functions and the outputs that they give.

In the following examples, we will run example data and in return, obtain
TASSEL association table reports in the form of an R `list` object containing
`tibble`-based R data frames.

## Calculate BLUEs
To caclulate best linear unbiased estimates (BLUEs), numeric phenotype data
can be used along with covariate and factor data *only* if it is intended to
control for field variation. Since genotype data is not needed for this
method, we can leave the `fitMarkers`, `kinship`, and `fastAssociation` to
`NULL` or `FALSE`:

```{r, echo=TRUE, eval=TRUE}
# Read in phenotype data
phenoPathCov <- system.file("extdata", "mdp_phenotype.txt", package = "rTASSEL")
tasPhenoCov <- rTASSEL::readPhenotypeFromPath(phenoPathCov)

# Calculate BLUEs
tasBLUE <- rTASSEL::assocModelFitter(
    tasObj = tasPhenoCov,
    formula = . ~ .,                  # <- All data is used!
    fitMarkers = FALSE,
    kinship = NULL,
    fastAssociation = FALSE
)

# Return BLUE output
str(tasBLUE)
```


## Calculate GLM
Similar to BLUEs, we can fit a generalized linear model (GLM) by simply
fitting marker data to our model. For this, we need a genotype data set
combined with our phenotype data in a `TasselGenotypePhenotype` class object:

```{r, echo=TRUE, eval=TRUE}
# Calculate GLM
tasGLM <- rTASSEL::assocModelFitter(
    tasObj = tasGenoPheno,             # <- our prior TASSEL object
    formula = list(EarHT, dpoll) ~ .,  # <- only EarHT and dpoll are ran
    fitMarkers = TRUE,                 # <- set this to TRUE for GLM
    kinship = NULL,
    fastAssociation = FALSE
)

# Return GLM output
str(tasGLM)
```


## Calculate MLM
Adding to our complexity, we can fit a mixed linear model (MLM) by adding
kinship to our analysis. In addition to the prior parameters, we will also
need a TASSEL kinship object (see *Create a kinship matrix object* in the
*Analysis - Relatedness* section):

```{r, echo=TRUE, eval=TRUE}
# Calculate MLM
tasMLM <- rTASSEL::assocModelFitter(
    tasObj = tasGenoPheno,             # <- our prior TASSEL object
    formula = EarHT ~ .,               # <- run only EarHT
    fitMarkers = TRUE,                 # <- set this to TRUE for GLM
    kinship = tasKin,                  # <- our prior kinship object
    fastAssociation = FALSE
)

# Return GLM output
str(tasMLM)
```


## Calculate Fast Association
Finally, we can run fast association analysis in our GLM model by setting
the `fastAssociation` parameter to `TRUE`. **NOTE**: this is only really
effective if you have many phenotype traits:

```{r, echo=TRUE, eval=TRUE}
# Read data - need only non missing data!
phenoPathFast <-system.file(
    "extdata",
    "mdp_traits_nomissing.txt",
    package = "rTASSEL"
)

# Creat rTASSEL object - use prior TASSEL genotype object
tasGenoPhenoFast <- rTASSEL::readGenotypePhenotype(
    genoPathOrObj = tasGenoHMP,
    phenoPathDFOrObj = phenoPathFast
)


# Calculate MLM
tasFAST <- rTASSEL::assocModelFitter(
    tasObj = tasGenoPhenoFast,         # <- our prior TASSEL object
    formula = . ~ .,                   # <- run all of the phenotype data
    fitMarkers = TRUE,                 # <- set this to TRUE for GLM
    kinship = NULL,
    fastAssociation = TRUE             # <- set this to TRUE for fast assoc.
)

# Return GLM output
str(tasFAST)
```



# Visualizations

## Manhattan plots
`rTASSEL` supports automated visualizations for Manhattan plots and linkage 
disequilibrium (LD) analyses. In this example, we will generate a Manhattan 
plot for our prior MLM object:

```{r, echo=TRUE, eval=FALSE}
# Generate Manhattan plot for ear height trait
manhattanEH <- manhattanPlot(
    assocStats = tasMLM$MLM_Stats,
    trait      = "EarHT",
    threshold  = 5
)
```


## LD plots
Similarly, we can also visualize LD using automated methods. Like most 
LD plots, it is wise to filter your genotype information to a specific region 
of interest:

```{r, echo=TRUE, eval=FALSE}
# Filter genotype table by position
tasGenoPhenoFilt <- filterGenotypeTableSites(
    tasObj              = tasGenoPheno,
    siteRangeFilterType = "position",
    startPos            = 228e6,
    endPos              = 300e6,
    startChr            = 2,
    endChr              = 2
)

# Generate and visualize LD
myLD <- ldPlot(
    tasObj  = tasGenoPhenoFilt,
    ldType  = "All",
    plotVal = "r2",
    verbose = FALSE
)
```



# Genomic Prediction

## Overview
`rTASSEL` also allows for phenotypic prediction through genotype information 
via genomic best linear unbiased predictors (gBLUPs). It proceeds by fitting a 
mixed model that uses kinship to capture covariance between taxa. The mixed 
model can calculate BLUPs for taxa that do not have phenotypes based on the 
phenotypes of lines with relationship information.

A phenotype dataset and a kinship matrix must be supplied as input to the 
method by selecting both then choosing Analysis/Genomic Selection. In addition 
to trait values, the phenotype dataset may also contain factors or covariates 
which will be used as fixed effects in the model. All taxa in the phenotype 
dataset can only appear once. No repeated values are allowed for a single 
taxon. When the analysis is run, the user is presented with the choice to run 
k-fold cross-validation. If cross- validation is selected, then the number of 
folds and the number of iterations can be entered. For each iteration and each 
fold within an iteration, the correlation between the observed and predicted 
values will be reported. If cross-validation is not selected, then the original 
observations, predicted values and PEVs (prediction error variance) will be 
reported for all taxa in the dataset.

When k-fold cross-validation is performed, only taxa with phenotypes and rows 
in the kinship matrix are used. That set of taxa are divided into k subsets of 
equal size. Each subset in turn is used as the validation set. Phenotypes of 
the individuals in the validation are set to 0 then predicted using the 
remaining individuals as the training set. The correlation (r) of the observed 
values and predicted values is calculated for the validation set and reported. 
The mean and standard deviation of the mean of the r's are calculated for each 
trait and reported in the comments section of the "Accuracy" data set that is 
output by the analysis. In general, the results are not very sensitive to the 
choice of k. The number of iterations affects the standard error of the mean 
for the accuracy estimates. The defaults of k = 5 and iterations = 20 will be 
adequate for most users.


```{r, echo=TRUE, eval=TRUE}
tasCV <- genomicPrediction(
    tasPhenoObj = tasGenoPheno,
    kinship     = tasKin,
    doCV        = TRUE,
    kFolds      = 5,
    nIter       = 1
)
head(tasCV)
```


