<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>R interface for TASSEL's association methods — assocModelFitter • rTASSEL</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="R interface for TASSEL's association methods — assocModelFitter"><meta property="og:description" content="This function acts as a front-end for TASSEL's extensive
   association analysis methods. Using this function, users can run
   the following TASSEL association methods:
best linear unbiased estimates (BLUEs)
generalized linear model (GLM)
mixed linear model
Fast association (Shabalin 2012)

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rTASSEL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.26</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/genotype_filtration.html">Filtering Genotype Tables</a>
    </li>
    <li>
      <a href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a>
    </li>
    <li>
      <a href="../articles/rtassel_walkthrough.html">Getting Started with rTASSEL</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R interface for TASSEL's association methods</h1>
    
    <div class="hidden name"><code>assocModelFitter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function acts as a front-end for TASSEL's extensive
   association analysis methods. Using this function, users can run
   the following TASSEL association methods:</p><ul><li><p>best linear unbiased estimates (BLUEs)</p></li>
<li><p>generalized linear model (GLM)</p></li>
<li><p>mixed linear model</p></li>
<li><p>Fast association (Shabalin 2012)</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">assocModelFitter</span><span class="op">(</span>
  <span class="va">tasObj</span>,
  <span class="va">formula</span>,
  fitMarkers <span class="op">=</span> <span class="cn">FALSE</span>,
  kinship <span class="op">=</span> <span class="cn">NULL</span>,
  fastAssociation <span class="op">=</span> <span class="cn">FALSE</span>,
  maxP <span class="op">=</span> <span class="fl">0.001</span>,
  maxThreads <span class="op">=</span> <span class="cn">NULL</span>,
  minClassSize <span class="op">=</span> <span class="fl">0</span>,
  outputFile <span class="op">=</span> <span class="cn">NULL</span>,
  biallelicOnly <span class="op">=</span> <span class="cn">FALSE</span>,
  appendAddDom <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>tasObj</dt>
<dd><p>An object of class <code>TasselGenotypePenotype</code>.</p></dd>


<dt>formula</dt>
<dd><p>An R-based linear model formula. The general layout of this
formula uses the following TASSEL data scheme:
<code>&lt;data&gt; ~ &lt;factor&gt; and/or &lt;covariate&gt;</code>. If all traits in a Phenotype
object should be ran, a simplified formula (<code>. ~ .</code>) can be used.
This scheme can also be used for running all <code>&lt;data&gt;</code> or
<code>&lt;factor&gt;</code> and/or <code>&lt;covariate&gt;</code> data as well. Single variables
are separated witha <code>+</code> operator. See vignette for further
clarification.</p></dd>


<dt>fitMarkers</dt>
<dd><p>Should marker data be fitted? If <code>TRUE</code>, GLM
analysis will be executed. If <code>FALSE</code>, BLUEs will be calculated.
Defaults to <code>FALSE</code>.</p></dd>


<dt>kinship</dt>
<dd><p>Should kinship data be accounted for in the model? If so,
a TASSEL kinship matrix object of class <code>TasselDistanceMatrix</code> must
be submitted. Defaults to <code>NULL</code>.</p></dd>


<dt>fastAssociation</dt>
<dd><p>Should TASSEL's Fast Association plugin be used?
Consider setting to <code>TRUE</code> if you have many phenotypes in your
data set.</p></dd>


<dt>maxP</dt>
<dd><p>Maximum p-value (0 - 1) to be reported. Currently works with
fast association only. Defaults to a p-value of <code>0.001</code>
will be used as a threshold. <strong>Note:</strong> p-value parameter will
not be used for BLUE analysis.</p></dd>


<dt>maxThreads</dt>
<dd><p>Maximum threads to be used when running fast association.
If <code>NULL</code>, all threads on machine will be used.</p></dd>


<dt>minClassSize</dt>
<dd><p>The minimum acceptable genotype class size. Genotypes
in a class with a smaller size will be set to missing. Defaults to 0.</p></dd>


<dt>outputFile</dt>
<dd><p>Output file prefix to be specified in case you want
to write data directly to disk. Highly recommended for large datasets.
If <code>NULL</code>, no data will be saved to disk. If a character</p></dd>


<dt>biallelicOnly</dt>
<dd><p>Only test sites that are bi-allelic. The alternative is
to test sites with two or more alleles. Defaults to <code>FALSE</code></p></dd>


<dt>appendAddDom</dt>
<dd><p>If true, additive and dominance effect estimates will
be added to the stats report for bi-allelic sites only. The effect will
only be estimated when the data source is genotype (not a probability).
The additive effect will always be non-negative. Defaults to <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns an R list containing <code>DataFrame</code>-based data frames</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brandon Monier, Terry Casstevens, Ed Buckler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

