<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>R interface for TASSEL's association methods — assocModelFitter • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="R interface for TASSEL's association methods — assocModelFitter"><meta name="description" content="This function acts as a front-end for TASSEL's extensive
   association analysis methods. Using this function, users can run
   the following TASSEL association methods:
best linear unbiased estimates (BLUEs)
generalized linear model (GLM)
mixed linear model
Fast association (Shabalin 2012)

"><meta property="og:description" content="This function acts as a front-end for TASSEL's extensive
   association analysis methods. Using this function, users can run
   the following TASSEL association methods:
best linear unbiased estimates (BLUEs)
generalized linear model (GLM)
mixed linear model
Fast association (Shabalin 2012)

"><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.11.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>R interface for TASSEL's association methods</h1>
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/v0.11.0/R/analysis_association_functions.R" class="external-link"><code>R/analysis_association_functions.R</code></a></small>
      <div class="d-none name"><code>assocModelFitter.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function acts as a front-end for TASSEL's extensive
   association analysis methods. Using this function, users can run
   the following TASSEL association methods:</p><ul><li><p>best linear unbiased estimates (BLUEs)</p></li>
<li><p>generalized linear model (GLM)</p></li>
<li><p>mixed linear model</p></li>
<li><p>Fast association (Shabalin 2012)</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assocModelFitter</span><span class="op">(</span></span>
<span>  <span class="va">tasObj</span>,</span>
<span>  <span class="va">formula</span>,</span>
<span>  fitMarkers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kinship <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fastAssociation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxP <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  maxThreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minClassSize <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  outputFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  biallelicOnly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  appendAddDom <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tasobj">tasObj<a class="anchor" aria-label="anchor" href="#arg-tasobj"></a></dt>
<dd><p>An object of class <code>TasselGenotypePenotype</code>.</p></dd>


<dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>An R-based linear model formula. The general layout of this
formula uses the following TASSEL data scheme:
<code>&lt;data&gt; ~ &lt;factor&gt; and/or &lt;covariate&gt;</code>. If all traits in a Phenotype
object should be ran, a simplified formula (<code>. ~ .</code>) can be used.
This scheme can also be used for running all <code>&lt;data&gt;</code> or
<code>&lt;factor&gt;</code> and/or <code>&lt;covariate&gt;</code> data as well. Single variables
are separated witha <code>+</code> operator. See vignette for further
clarification.</p></dd>


<dt id="arg-fitmarkers">fitMarkers<a class="anchor" aria-label="anchor" href="#arg-fitmarkers"></a></dt>
<dd><p>Should marker data be fitted? If <code>TRUE</code>, GLM
analysis will be executed. If <code>FALSE</code>, BLUEs will be calculated.
Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg-kinship">kinship<a class="anchor" aria-label="anchor" href="#arg-kinship"></a></dt>
<dd><p>Should kinship data be accounted for in the model? If so,
a TASSEL kinship matrix object of class <code>TasselDistanceMatrix</code> must
be submitted. Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-fastassociation">fastAssociation<a class="anchor" aria-label="anchor" href="#arg-fastassociation"></a></dt>
<dd><p>Should TASSEL's Fast Association plugin be used?
Consider setting to <code>TRUE</code> if you have many phenotypes in your
data set.</p></dd>


<dt id="arg-maxp">maxP<a class="anchor" aria-label="anchor" href="#arg-maxp"></a></dt>
<dd><p>Maximum p-value (0 - 1) to be reported. Currently works with
fast association only. Defaults to a p-value of <code>0.001</code>
will be used as a threshold. <strong>Note:</strong> p-value parameter will
not be used for BLUE analysis.</p></dd>


<dt id="arg-maxthreads">maxThreads<a class="anchor" aria-label="anchor" href="#arg-maxthreads"></a></dt>
<dd><p>Maximum threads to be used when running fast association.
If <code>NULL</code>, all threads on machine will be used.</p></dd>


<dt id="arg-minclasssize">minClassSize<a class="anchor" aria-label="anchor" href="#arg-minclasssize"></a></dt>
<dd><p>The minimum acceptable genotype class size. Genotypes
in a class with a smaller size will be set to missing. Defaults to 0.</p></dd>


<dt id="arg-outputfile">outputFile<a class="anchor" aria-label="anchor" href="#arg-outputfile"></a></dt>
<dd><p>Output file prefix to be specified in case you want
to write data directly to disk. Highly recommended for large datasets.
If <code>NULL</code>, no data will be saved to disk. If a character</p></dd>


<dt id="arg-bialleliconly">biallelicOnly<a class="anchor" aria-label="anchor" href="#arg-bialleliconly"></a></dt>
<dd><p>Only test sites that are bi-allelic. The alternative is
to test sites with two or more alleles. Defaults to <code>FALSE</code></p></dd>


<dt id="arg-appendadddom">appendAddDom<a class="anchor" aria-label="anchor" href="#arg-appendadddom"></a></dt>
<dd><p>If true, additive and dominance effect estimates will
be added to the stats report for bi-allelic sites only. The effect will
only be estimated when the data source is genotype (not a probability).
The additive effect will always be non-negative. Defaults to <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns an R list containing <code>DataFrame</code>-based data frames</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

