<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filter genotype table by sites — filterGenotypeTableSites • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Filter genotype table by sites — filterGenotypeTableSites"><meta name="description" content="This function will filter R objects of
   TasselGenotypePhenotype class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   FilterSiteBuilder plugin."><meta property="og:description" content="This function will filter R objects of
   TasselGenotypePhenotype class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   FilterSiteBuilder plugin."><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Filter genotype table by sites</h1>
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/master/R/filter_functions.R" class="external-link"><code>R/filter_functions.R</code></a></small>
      <div class="d-none name"><code>filterGenotypeTableSites.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will filter R objects of
   <code>TasselGenotypePhenotype</code> class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   <code>FilterSiteBuilder</code> plugin.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filterGenotypeTableSites</span><span class="op">(</span></span>
<span>  <span class="va">tasObj</span>,</span>
<span>  siteMinCount <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  siteMinAlleleFreq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  siteMaxAlleleFreq <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minHeterozygous <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxHeterozygous <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  removeMinorSNPStates <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  removeSitesWithIndels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  siteRangeFilterType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"sites"</span>, <span class="st">"position"</span><span class="op">)</span>,</span>
<span>  startSite <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endSite <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  startChr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  startPos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endChr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endPos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gRangesObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrPosFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedFile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tasobj">tasObj<a class="anchor" aria-label="anchor" href="#arg-tasobj"></a></dt>
<dd><p>An object of class <code>TasselGenotypePenotype</code>.</p></dd>


<dt id="arg-sitemincount">siteMinCount<a class="anchor" aria-label="anchor" href="#arg-sitemincount"></a></dt>
<dd><p>Site minimum count of alleles not unknown. Can range
from 0 to inf. Defaults to 0.</p></dd>


<dt id="arg-siteminallelefreq">siteMinAlleleFreq<a class="anchor" aria-label="anchor" href="#arg-siteminallelefreq"></a></dt>
<dd><p>Site minimum minor allele frequency. Can range
from 0 to 1.0. Defaults to 0.0.</p></dd>


<dt id="arg-sitemaxallelefreq">siteMaxAlleleFreq<a class="anchor" aria-label="anchor" href="#arg-sitemaxallelefreq"></a></dt>
<dd><p>Site maximum minor allele frequency. Can range
from 0 to 1.0. Defaults to 1.0.</p></dd>


<dt id="arg-minheterozygous">minHeterozygous<a class="anchor" aria-label="anchor" href="#arg-minheterozygous"></a></dt>
<dd><p>Min heterozygous proportion. Can range from 0 to 1.0.
Defaults to 0.0.</p></dd>


<dt id="arg-maxheterozygous">maxHeterozygous<a class="anchor" aria-label="anchor" href="#arg-maxheterozygous"></a></dt>
<dd><p>Max heterozygous proportion. Can range from 0 to 1.0.
Defaults to 1.0.</p></dd>


<dt id="arg-removeminorsnpstates">removeMinorSNPStates<a class="anchor" aria-label="anchor" href="#arg-removeminorsnpstates"></a></dt>
<dd><p>Remove minor SNP states. Defaults to
<code>FALSE</code>.</p></dd>


<dt id="arg-removesiteswithindels">removeSitesWithIndels<a class="anchor" aria-label="anchor" href="#arg-removesiteswithindels"></a></dt>
<dd><p>Remove sites containing an indel
(<code>+</code> or <code>-</code>). Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg-siterangefiltertype">siteRangeFilterType<a class="anchor" aria-label="anchor" href="#arg-siterangefiltertype"></a></dt>
<dd><p>True if filtering by site numbers. False if
filtering by chromosome and position. Options are
<code>none</code>, <code>sites</code>, or <code>position</code>. Defaults to <code>none</code>.</p></dd>


<dt id="arg-startsite">startSite<a class="anchor" aria-label="anchor" href="#arg-startsite"></a></dt>
<dd><p>The start site. Defaults to 0.</p></dd>


<dt id="arg-endsite">endSite<a class="anchor" aria-label="anchor" href="#arg-endsite"></a></dt>
<dd><p>The end site. Defaults to 0.</p></dd>


<dt id="arg-startchr">startChr<a class="anchor" aria-label="anchor" href="#arg-startchr"></a></dt>
<dd><p>Start chromosome for site filtration range if <code>position</code>
is chosen from <code>siteRangeFilterType</code>. Needs end chromosome
(<code>endChr</code>) to work.</p></dd>


<dt id="arg-startpos">startPos<a class="anchor" aria-label="anchor" href="#arg-startpos"></a></dt>
<dd><p>Physical start position (bp) for filtration range if
<code>position</code> is chosen from <code>siteRangeFilterType</code>. If
<code>NULL</code>, the first physical position in the data set will be
chosen.</p></dd>


<dt id="arg-endchr">endChr<a class="anchor" aria-label="anchor" href="#arg-endchr"></a></dt>
<dd><p>End chromosome for site filtration range if <code>position</code>
is chosen from <code>siteRangeFilterType</code>. Needs start chromosome
(<code>endChr</code>) to work.</p></dd>


<dt id="arg-endpos">endPos<a class="anchor" aria-label="anchor" href="#arg-endpos"></a></dt>
<dd><p>Physical end position (bp) for filtration range if
<code>position</code> is chosen from <code>siteRangeFilterType</code>. If
<code>NULL</code>, the last physical position in the data set will be
chosen.</p></dd>


<dt id="arg-grangesobj">gRangesObj<a class="anchor" aria-label="anchor" href="#arg-grangesobj"></a></dt>
<dd><p>Filter genotype table by <code>GenomicRanges</code> object.
If this parameter is selected, you cannot utilize the parameters,
<code>chrPosFile</code> or <code>bedFile</code>. Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-chrposfile">chrPosFile<a class="anchor" aria-label="anchor" href="#arg-chrposfile"></a></dt>
<dd><p>An optional chromosome position file path of
<code>character</code> class. Defaults to <code>NULL</code>. <strong>Note:</strong>
a chromosome position file must contain correct formatting
(e.g. a two column file with the header of
<code>c("Chromosome", "Position")</code>).</p></dd>


<dt id="arg-bedfile">bedFile<a class="anchor" aria-label="anchor" href="#arg-bedfile"></a></dt>
<dd><p>An optional BED coordinate file path of
<code>character</code> class. Defaults to <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns an object of <code>TasselGenotypePhenotype</code> class.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

