<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function will filter R objects of
   TasselGenotypePhenotype class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   FilterSiteBuilder plugin."><title>Filter genotype table by sites — filterGenotypeTableSites • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filter genotype table by sites — filterGenotypeTableSites"><meta property="og:description" content="This function will filter R objects of
   TasselGenotypePhenotype class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   FilterSiteBuilder plugin."><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.31</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a>
    <a class="dropdown-item" href="../articles/rtassel_walkthrough.html">Getting Started with rTASSEL</a>
    <a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a>
    <a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Filter genotype table by sites</h1>
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/HEAD/R/filter_functions.R" class="external-link"><code>R/filter_functions.R</code></a></small>
      <div class="d-none name"><code>filterGenotypeTableSites.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will filter R objects of
   <code>TasselGenotypePhenotype</code> class containing genotype tables.
   The parameters for this function are derived from TASSEL's
   <code>FilterSiteBuilder</code> plugin.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filterGenotypeTableSites</span><span class="op">(</span></span>
<span>  <span class="va">tasObj</span>,</span>
<span>  siteMinCount <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  siteMinAlleleFreq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  siteMaxAlleleFreq <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minHeterozygous <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxHeterozygous <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  removeMinorSNPStates <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  removeSitesWithIndels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  siteRangeFilterType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"sites"</span>, <span class="st">"position"</span><span class="op">)</span>,</span>
<span>  startSite <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endSite <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  startChr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  startPos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endChr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  endPos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gRangesObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrPosFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedFile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tasObj</dt>
<dd><p>An object of class <code>TasselGenotypePenotype</code>.</p></dd>


<dt>siteMinCount</dt>
<dd><p>Site minimum count of alleles not unknown. Can range
from 0 to inf. Defaults to 0.</p></dd>


<dt>siteMinAlleleFreq</dt>
<dd><p>Site minimum minor allele frequency. Can range
from 0 to 1.0. Defaults to 0.0.</p></dd>


<dt>siteMaxAlleleFreq</dt>
<dd><p>Site maximum minor allele frequency. Can range
from 0 to 1.0. Defaults to 1.0.</p></dd>


<dt>minHeterozygous</dt>
<dd><p>Min heterozygous proportion. Can range from 0 to 1.0.
Defaults to 0.0.</p></dd>


<dt>maxHeterozygous</dt>
<dd><p>Max heterozygous proportion. Can range from 0 to 1.0.
Defaults to 1.0.</p></dd>


<dt>removeMinorSNPStates</dt>
<dd><p>Remove minor SNP states. Defaults to
<code>FALSE</code>.</p></dd>


<dt>removeSitesWithIndels</dt>
<dd><p>Remove sites containing an indel
(<code>+</code> or <code>-</code>). Defaults to <code>FALSE</code>.</p></dd>


<dt>siteRangeFilterType</dt>
<dd><p>True if filtering by site numbers. False if
filtering by chromosome and position. Options are
<code>none</code>, <code>sites</code>, or <code>position</code>. Defaults to <code>none</code>.</p></dd>


<dt>startSite</dt>
<dd><p>The start site. Defaults to 0.</p></dd>


<dt>endSite</dt>
<dd><p>The end site. Defaults to 0.</p></dd>


<dt>startChr</dt>
<dd><p>Start chromosome for site filtration range if <code>position</code>
is chosen from <code>siteRangeFilterType</code>. Needs end chromosome
(<code>endChr</code>) to work.</p></dd>


<dt>startPos</dt>
<dd><p>Physical start position (bp) for filtration range if
<code>position</code> is chosen from <code>siteRangeFilterType</code>. If
<code>NULL</code>, the first physical position in the data set will be
chosen.</p></dd>


<dt>endChr</dt>
<dd><p>End chromosome for site filtration range if <code>position</code>
is chosen from <code>siteRangeFilterType</code>. Needs start chromosome
(<code>endChr</code>) to work.</p></dd>


<dt>endPos</dt>
<dd><p>Physical end position (bp) for filtration range if
<code>position</code> is chosen from <code>siteRangeFilterType</code>. If
<code>NULL</code>, the last physical position in the data set will be
chosen.</p></dd>


<dt>gRangesObj</dt>
<dd><p>Filter genotype table by <code>GenomicRanges</code> object.
If this parameter is selected, you cannot utilize the parameters,
<code>chrPosFile</code> or <code>bedFile</code>. Defaults to <code>NULL</code>.</p></dd>


<dt>chrPosFile</dt>
<dd><p>An optional chromosome position file path of
<code>character</code> class. Defaults to <code>NULL</code>. <strong>Note:</strong>
a chromosome position file must contain correct formatting
(e.g. a two column file with the header of
<code>c("Chromosome", "Position")</code>).</p></dd>


<dt>bedFile</dt>
<dd><p>An optional BED coordinate file path of
<code>character</code> class. Defaults to <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns an object of <code>TasselGenotypePhenotype</code> class.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

