<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.11.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="AssociationResults-class.html">show(<i>&lt;AssociationResults&gt;</i>)</a></code>

        </dt>
        <dd>AssociationResults Class</dd>

        <dt>

          <code><a href="AssociationResults-validity.html">AssociationResults-validity</a></code>

        </dt>
        <dd>AssociationResults validation</dd>

        <dt>

          <code><a href="AssociationResultsBLUE-class.html">AssociationResultsBLUE-class</a></code>

        </dt>
        <dd>AssociationResultsBLUE Class</dd>

        <dt>

          <code><a href="AssociationResultsFast-class.html">AssociationResultsFast-class</a></code>

        </dt>
        <dd>AssociationResultsFast Class</dd>

        <dt>

          <code><a href="AssociationResultsGLM-class.html">AssociationResultsGLM-class</a></code>

        </dt>
        <dd>AssociationResultsGLM Class</dd>

        <dt>

          <code><a href="AssociationResultsMLM-class.html">AssociationResultsMLM-class</a></code>

        </dt>
        <dd>AssociationResultsMLM Class</dd>

        <dt>

          <code><a href="AssociationResultsStepwise-class.html">AssociationResultsStepwise-class</a></code>

        </dt>
        <dd>AssociationResultsStepwise Class</dd>

        <dt>

          <code><a href="PCAResults-class.html">show(<i>&lt;PCAResults&gt;</i>)</a></code>

        </dt>
        <dd>PCAResults Class</dd>

        <dt>

          <code><a href="PCAResults-validity.html">PCAResults-validity</a></code>

        </dt>
        <dd>PCAResults validation</dd>

        <dt>

          <code><a href="TasselDistanceMatrix-class.html">show(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>TasselDistanceMatrix Class</dd>

        <dt>

          <code><a href="TasselGenotype-class.html">TasselGenotype-class</a></code>

        </dt>
        <dd>TasselGenotype Class</dd>

        <dt>

          <code><a href="TasselGenotypePhenotype-class.html">show(<i>&lt;TasselGenotypePhenotype&gt;</i>)</a></code>

        </dt>
        <dd>TasselGenotypePhenotype Class</dd>

        <dt>

          <code><a href="TasselNumericGenotype-class.html">show(<i>&lt;TasselNumericGenotype&gt;</i>)</a></code>

        </dt>
        <dd>TasselNumericGenotype Class Definition</dd>

        <dt>

          <code><a href="TasselPhenotype-class.html">TasselPhenotype-class</a></code>

        </dt>
        <dd>TasselPhenotype Class Definition</dd>

        <dt>

          <code><a href="as.matrix.TasselDistanceMatrix.html">as.matrix(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Coerce matrix from TasselDistanceMatrix class</dd>

        <dt>

          <code><a href="as.matrix.TasselGenotypePhenotype.html">as.matrix(<i>&lt;TasselGenotypePhenotype&gt;</i>)</a></code>

        </dt>
        <dd>Coerce genotype table to R matrix</dd>

        <dt>

          <code><a href="asTasselDistanceMatrix.html">asTasselDistanceMatrix()</a></code>

        </dt>
        <dd>Coerce matrix to TasselDistanceMatrix object</dd>

        <dt>

          <code><a href="assocModelFitter.html">assocModelFitter()</a></code>

        </dt>
        <dd>R interface for TASSEL's association methods</dd>

        <dt>

          <code><a href="associationType.html">associationType()</a></code>

        </dt>
        <dd>Return GWAS association type</dd>

        <dt>

          <code><a href="attributeData.html">attributeData()</a></code>

        </dt>
        <dd>Return TASSEL attribute data</dd>

        <dt>

          <code><a href="colnames-TasselDistanceMatrix-method.html">colnames(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Column names</dd>

        <dt>

          <code><a href="concatenate.html">concatenate()</a></code>

        </dt>
        <dd>Concatenate phenotype tables</dd>

        <dt>

          <code><a href="createTree.html">createTree()</a></code>

        </dt>
        <dd>R interface for TASSEL's tree creation methods</dd>

        <dt>

          <code><a href="dim-TasselDistanceMatrix-method.html">dim(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Get dimensions of TasselDistanceMatrix object</dd>

        <dt>

          <code><a href="distanceMatrix.html">distanceMatrix()</a></code>

        </dt>
        <dd>Create a TASSEL distance matrix</dd>

        <dt>

          <code><a href="exportGenotypeTable.html">exportGenotypeTable()</a></code>

        </dt>
        <dd>Export Genotype Table to Disk</dd>

        <dt>

          <code><a href="filterGenotypeTableBySiteName.html">filterGenotypeTableBySiteName()</a></code>

        </dt>
        <dd>Filter genotype table by site IDs</dd>

        <dt>

          <code><a href="filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>

        </dt>
        <dd>Filter genotype table by sites</dd>

        <dt>

          <code><a href="filterGenotypeTableTaxa.html">filterGenotypeTableTaxa()</a></code>

        </dt>
        <dd>Filter genotype table by taxa</dd>

        <dt>

          <code><a href="genomicPrediction.html">genomicPrediction()</a></code>

        </dt>
        <dd>R interface for TASSEL's genomic prediction capabilities</dd>

        <dt>

          <code><a href="getPhenotypeDF.html">getPhenotypeDF()</a></code>

        </dt>
        <dd>Get an R/<code>DataFrame</code> phenotype data frame from TASSEL object</dd>

        <dt>

          <code><a href="getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable()</a></code>

        </dt>
        <dd>Create Summarized Experiment from a TASSEL Genotype Table</dd>

        <dt>

          <code><a href="imputeLDKNNi.html">imputeLDKNNi()</a></code>

        </dt>
        <dd>LD KNNi imputation</dd>

        <dt>

          <code><a href="imputeNumeric.html">imputeNumeric()</a></code>

        </dt>
        <dd>Imputation methods in Numerical Transformations</dd>

        <dt>

          <code><a href="intersectJoin.html">intersectJoin()</a></code>

        </dt>
        <dd>Intersect join phenotype tables</dd>

        <dt>

          <code><a href="javaRefObj.html">javaRefObj()</a></code>

        </dt>
        <dd>Return <code>rJava</code> reference object</dd>

        <dt>

          <code><a href="kinshipMatrix.html">kinshipMatrix()</a></code>

        </dt>
        <dd>Create a TASSEL kinship matrix</dd>

        <dt>

          <code><a href="ldJavaApp.html">ldJavaApp()</a></code>

        </dt>
        <dd>Linkage desequilibrium visualization application</dd>

        <dt>

          <code><a href="ldPlot.html">ldPlot()</a></code>

        </dt>
        <dd>Linkage disequilibrium plot</dd>

        <dt>

          <code><a href="linkageDiseq.html">linkageDiseq()</a></code>

        </dt>
        <dd>Calculate linkage disequilibrium from an rTASSEL genotype dataset.</dd>

        <dt>

          <code><a href="mds.html">mds()</a></code>

        </dt>
        <dd>Run MDS on <code>TasselDistanceMatrix</code> objects</dd>

        <dt>

          <code><a href="mergeGenotypeTables.html">mergeGenotypeTables()</a></code>

        </dt>
        <dd>Merge genotype tables</dd>

        <dt>

          <code><a href="ncol-TasselDistanceMatrix-method.html">ncol(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Number of columns</dd>

        <dt>

          <code><a href="nrow-TasselDistanceMatrix-method.html">nrow(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Number of rows</dd>

        <dt>

          <code><a href="pca.html">pca()</a></code>

        </dt>
        <dd>Run PCA on Genotype Table</dd>

        <dt>

          <code><a href="plotManhattan.html">plotManhattan()</a></code>

        </dt>
        <dd>Create a Manhattan plot from rTASSEL association output</dd>

        <dt>

          <code><a href="plotManhattanQC.html">plotManhattanQC()</a></code>

        </dt>
        <dd>Create a QC Manhattan plots from rTASSEL association output</dd>

        <dt>

          <code><a href="plotPCA.html">plotPCA()</a></code>

        </dt>
        <dd>Generate PCA plot</dd>

        <dt>

          <code><a href="plotQQ.html">plotQQ()</a></code>

        </dt>
        <dd>Create a QQ plot from rTASSEL association output</dd>

        <dt>

          <code><a href="plotScree.html">plotScree()</a></code>

        </dt>
        <dd>Generate scree plots</dd>

        <dt>

          <code><a href="positionList.html">positionList()</a></code>

        </dt>
        <dd>Get position list metadata from genotype table</dd>

        <dt>

          <code><a href="readGenotype.html">readGenotype()</a></code>

        </dt>
        <dd>Read Genotype Data</dd>

        <dt>

          <code><a href="readGenotypePhenotype.html">readGenotypePhenotype()</a></code>

        </dt>
        <dd>Wrapper function of TasselGenotypePhenotype class for GenotypePhenotype combined data</dd>

        <dt>

          <code><a href="readGenotypeTableFromGigwa.html">readGenotypeTableFromGigwa()</a></code>

        </dt>
        <dd>Read genotype data from GIGWA using QBMS</dd>

        <dt>

          <code><a href="readGenotypeTableFromPath.html">readGenotypeTableFromPath()</a></code>

        </dt>
        <dd>Wrapper function of TasselGenotypePhenotype class for genotype data</dd>

        <dt>

          <code><a href="readNumericGenotypeFromRMatrix.html">readNumericGenotypeFromRMatrix()</a></code>

        </dt>
        <dd>Read Genotype Data from R Matrix</dd>

        <dt>

          <code><a href="readPhenotype.html">readPhenotype()</a></code>

        </dt>
        <dd>Read and convert phenotype data into TASSEL 5 phenotype objects</dd>

        <dt>

          <code><a href="readPhenotypeFromDataFrame.html">readPhenotypeFromDataFrame()</a></code>

        </dt>
        <dd>Wrapper function of TasselGenotypePhenotype class for phenotype data from an R data frame</dd>

        <dt>

          <code><a href="readPhenotypeFromPath.html">readPhenotypeFromPath()</a></code>

        </dt>
        <dd>Wrapper function of TasselGenotypePhenotype class for phenotype data from a path.</dd>

        <dt>

          <code><a href="readTasselDistanceMatrix.html">readTasselDistanceMatrix()</a></code>

        </dt>
        <dd>read TASSEL distance matrix object from file</dd>

        <dt>

          <code><a href="reportNames.html">reportNames()</a></code> <code><a href="reportNames.html">traitNames(<i>&lt;AssociationResults&gt;</i>)</a></code>

        </dt>
        <dd>Return report names</dd>

        <dt>

          <code><a href="rownames-TasselDistanceMatrix-method.html">rownames(<i>&lt;TasselDistanceMatrix&gt;</i>)</a></code>

        </dt>
        <dd>Row names</dd>

        <dt>

          <code><a href="seqDiversity.html">seqDiversity()</a></code>

        </dt>
        <dd>Calculate sequence diversity</dd>

        <dt>

          <code><a href="setupTASSEL.html">setupTASSEL()</a></code>

        </dt>
        <dd>Download and configure TASSEL JAR files from Maven Central</dd>

        <dt>

          <code><a href="show-TasselGenotype-method.html">show(<i>&lt;TasselGenotype&gt;</i>)</a></code>

        </dt>
        <dd>Display TasselGenotype Object</dd>

        <dt>

          <code><a href="show-TasselPhenotype-method.html">show(<i>&lt;TasselPhenotype&gt;</i>)</a></code>

        </dt>
        <dd>Display summary information of a TasselPhenotype object</dd>

        <dt>

          <code><a href="siteSummary.html">siteSummary()</a></code>

        </dt>
        <dd>Get site summary of genotype table</dd>

        <dt>

          <code><a href="startLogger.html">startLogger()</a></code>

        </dt>
        <dd>Start TASSEL logging information</dd>

        <dt>

          <code><a href="stepwiseModelFitter.html">stepwiseModelFitter()</a></code>

        </dt>
        <dd>Stepwise Model Fitter</dd>

        <dt>

          <code><a href="tableReport.html">tableReport()</a></code>

        </dt>
        <dd>Return selected table report</dd>

        <dt>

          <code><a href="taxaList.html">taxaList()</a></code>

        </dt>
        <dd>Get list of taxa from TASSEL data</dd>

        <dt>

          <code><a href="taxaSummary.html">taxaSummary()</a></code>

        </dt>
        <dd>Get taxa summary of genotype table</dd>

        <dt>

          <code><a href="traitNames.html">traitNames()</a></code>

        </dt>
        <dd>Return trait names</dd>

        <dt>

          <code><a href="treeJavaApp.html">treeJavaApp()</a></code>

        </dt>
        <dd>R interface for Archaeopteryx interactive tree viewer</dd>

        <dt>

          <code><a href="unionJoin.html">unionJoin()</a></code>

        </dt>
        <dd>Union join phenotype tables</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

