<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a QC Manhattan plots from rTASSEL association output — plotManhattanQC • rTASSEL</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create a QC Manhattan plots from rTASSEL association output — plotManhattanQC"><meta name="description" content='This function allows for quick generation of a QC
   plot from rTASSEL association statistical output data. The main goal
   of this function is to provide "zoomed" in Manhattan plots that
   typically fall within genomic ranges of interest plus a flanking
   window up- and downstream of the ranges.'><meta property="og:description" content='This function allows for quick generation of a QC
   plot from rTASSEL association statistical output data. The main goal
   of this function is to provide "zoomed" in Manhattan plots that
   typically fall within genomic ranges of interest plus a flanking
   window up- and downstream of the ranges.'><meta property="og:image" content="https://rtassel.maizegenetics.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Create a QC Manhattan plots from rTASSEL association output</h1>
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/master/R/plot_manhattan_qc.R" class="external-link"><code>R/plot_manhattan_qc.R</code></a></small>
      <div class="d-none name"><code>plotManhattanQC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function allows for quick generation of a QC
   plot from rTASSEL association statistical output data. The main goal
   of this function is to provide "zoomed" in Manhattan plots that
   typically fall within genomic ranges of interest plus a flanking
   window up- and downstream of the ranges.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotManhattanQC</span><span class="op">(</span></span>
<span>  <span class="va">assocRes</span>,</span>
<span>  trait <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">gr</span>,</span>
<span>  window <span class="op">=</span> <span class="fl">1e+05</span>,</span>
<span>  threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  classicNames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-assocres">assocRes<a class="anchor" aria-label="anchor" href="#arg-assocres"></a></dt>
<dd><p>An object of type <code>AssociationResults</code></p></dd>


<dt id="arg-trait">trait<a class="anchor" aria-label="anchor" href="#arg-trait"></a></dt>
<dd><p>Which phenotypic trait do you want to plot? If set to
<code>NULL</code>, this will generate a faceted plot with all mapped traits.</p></dd>


<dt id="arg-gr">gr<a class="anchor" aria-label="anchor" href="#arg-gr"></a></dt>
<dd><p>Genomic ranges of interest. Can be passed as a <code>GRanges</code>
object or a <code>data.frame</code> object.</p></dd>


<dt id="arg-window">window<a class="anchor" aria-label="anchor" href="#arg-window"></a></dt>
<dd><p>Window size (base-pairs) flanking surround reference range.
Defaults to <code>100000</code> (100,000 base-pairs).</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>User-defined \(-log_{10}(p)\)-value threshold for
significant marker determination. Once specified any marker points
higher than this line will be highlighted.</p></dd>


<dt id="arg-classicnames">classicNames<a class="anchor" aria-label="anchor" href="#arg-classicnames"></a></dt>
<dd><p>Do you want to plot classical gene names instead?
NOTE: this will need a <code>classical_id</code> column for the "ranges of
interest" data. Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg-interactive">interactive<a class="anchor" aria-label="anchor" href="#arg-interactive"></a></dt>
<dd><p>Do you want to produce an interactive visualization?
Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Should messages be printed to console? Defaults to
<code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a <code>ggplot2</code> object</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

