<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Filtering Genotype Tables • rTASSEL</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet">
<link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet">
<link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Filtering Genotype Tables">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Filtering Genotype Tables</h1>
                        <h4 data-toc-skip class="author">Brandon
Monier</h4>
            
            <h4 data-toc-skip class="date">2025-10-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/master/vignettes/genotype_filtration.Rmd" class="external-link"><code>vignettes/genotype_filtration.Rmd</code></a></small>
      <div class="d-none name"><code>genotype_filtration.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="genotype-table-overview">Genotype table overview<a class="anchor" aria-label="anchor" href="#genotype-table-overview"></a>
</h3>
<p>In this vignette, we will discuss how you can filter genotype tables
in <code>rTASSEL</code>. As a reminder, a genotype table in TASSEL is
essentially an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">n \times p</annotation></semantics></math>
matrix consisting of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
samples or taxa with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
genotypic observations. To illustrate this, see the diagram below:</p>
<p><img src="../reference/figures/genotype_table_diagram.png"></p>
<p>In this illustration, sequence information is processed for variant
sites of each parental copy. Variant data is partitioned into an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">n \times p</annotation></semantics></math>
matrix for each sample and relevant variant marker information
(e.g. <code>0:2</code>). This includes:</p>
<ul>
<li>The TASSEL index for each variant site
(<code>&lt;this value&gt; :</code>)
<ul>
<li>
<em>Note:</em> since TASSEL is coded in Java, this value will start
at 0!</li>
</ul>
</li>
<li>The physical position (bp) for each variant site
(<code>: &lt;this value&gt;</code>)</li>
</ul>
<p>Genotype information (e.g. each
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">n \times p</annotation></semantics></math>
element) can be viewed in several ways. The default view (as shown
above) reflects the following:</p>
<table class="table">
<colgroup>
<col width="18%">
<col width="66%">
<col width="15%">
</colgroup>
<thead><tr class="header">
<th>Cell Type</th>
<th>Definition</th>
<th>Color</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Major alleles</td>
<td>The most frequently observed allele of each column</td>
<td>Yellow</td>
</tr>
<tr class="even">
<td>Minor alleles</td>
<td>The least frequently observed allele of each column</td>
<td>Blue</td>
</tr>
<tr class="odd">
<td>Heterozygous</td>
<td>Sites where parental copies differ (e.g. <code>C:G</code>)</td>
<td>Green</td>
</tr>
<tr class="even">
<td>Unknown</td>
<td>Sites where no information is given (e.g. <code>NA</code>s)</td>
<td>Grey/White</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="filtering-basics">Filtering basics<a class="anchor" aria-label="anchor" href="#filtering-basics"></a>
</h3>
<p>Since TASSEL allows for many features to be represented in a genotype
table, <code>rTASSEL</code>’s filtration parameters are numerous. In the
following sections, we will discuss how each set of parameters work. In
general, these filtration schemes can be applied for samples (i.e. rows)
and marker information (i.e. columns).</p>
<p>The two main methods to accomplish this are:</p>
<ul>
<li><code><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa()</a></code></li>
<li><code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code></li>
</ul>
<p>For each section example code will be given with an illustrated
genotype table to explain each method. Prior knowledge about
<code>rTASSEL</code> data structures is needed so please refere to the
vignette <em>Getting Started with rTASSEL</em> if you are new to this
package!</p>
<p><em>Note:</em> We will be using the the base R pipe operator
<code>|&gt;</code> for the following examples. More information about
this can be found <a href="https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="filtering-by-sites">Filtering by sites…<a class="anchor" aria-label="anchor" href="#filtering-by-sites"></a>
</h2>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>In order to filter by variant sites (e.g. marker information), we can
use the function <code><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites()</a></code>. This function
was modeled after TASSEL’s <code>FilterSiteBuilder</code> plugin to
ensure familiarity between the two methods. The parameters can be broken
up into several components:</p>
<ul>
<li>Genotype information
<ul>
<li><code>siteMinCount</code></li>
<li><code>siteMinAlleleFreq</code></li>
<li><code>siteMaxAlleleFreq</code></li>
<li><code>minHeterozygous</code></li>
<li><code>maxHeterozygous</code></li>
</ul>
</li>
<li>Indexed variant sites
<ul>
<li><code>startSite</code></li>
<li><code>endSite</code></li>
</ul>
</li>
<li>Physical marker positions
<ul>
<li><code>startChr</code></li>
<li><code>endChr</code></li>
<li><code>startPos</code></li>
<li><code>endPos</code></li>
</ul>
</li>
<li>R objects
<ul>
<li><code>gRangesObj</code></li>
</ul>
</li>
<li>External files
<ul>
<li><code>bedFile</code></li>
<li><code>chrPosFile</code></li>
</ul>
</li>
</ul>
<p><em>Note:</em> While the following examples will illustrate these
parameters in simplified formats, genotype information parameters
<em>and</em> marker index and position parameters (i.e. indexed variant
sites, physical marker positions, R objects, or external files) can be
passed simultaneously!</p>
<p>Let’s assume we have an <code>rTASSEL</code> data object that
contains a genotype table:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span></span></code></pre></div>
<pre><code><span><span class="co">## A TasselGenotypePhenotype Dataset</span></span>
<span><span class="co">##   Class.............. TasselGenotypePhenotype </span></span>
<span><span class="co">##   Taxa............... 5 </span></span>
<span><span class="co">##   Positions.......... 5</span></span>
<span><span class="co">##   Taxa x Positions... 25</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">##   Genotype Table..... [x]</span></span>
<span><span class="co">##   Phenotype Table.... [ ]</span></span></code></pre>
<p>Let’s also assume it contains the same structure as the prior
illustration:</p>
<p><img src="../reference/figures/filter_site_01.png"></p>
<p>In order for us to understand how the next few parameters work, we
will illustrate how allelic frequencies are calculated:</p>
<p><img src="../reference/figures/filter_site_02.png"></p>
</div>
<div class="section level3">
<h3 id="using-allelic-frequency">…using allelic frequency<a class="anchor" aria-label="anchor" href="#using-allelic-frequency"></a>
</h3>
<p>If we want to filter by minor allele frequency, we can use the
following parameters…</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteMinAlleleFreq <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>        siteMaxAlleleFreq <span class="op">=</span> <span class="fl">1.0</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>…which will remove the following columns:</p>
<p><img src="../reference/figures/filter_site_03.png"></p>
<p>We can also set a range using a limit with the
<code>siteMaxAlleleFreq</code> parameter…</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteMinAlleleFreq <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>        siteMaxAlleleFreq <span class="op">=</span> <span class="fl">0.3</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>…which will remove the following columns:</p>
<p><img src="../reference/figures/filter_site_04.png"></p>
</div>
<div class="section level3">
<h3 id="using-na-data">…using <code>NA</code> data<a class="anchor" aria-label="anchor" href="#using-na-data"></a>
</h3>
<p>We can also filter by missing sites (encoded in TASSEL with
<code>N</code>). Similar to how we calculated frequency of minor
alleles, we can apply the same logic with missing data with the
<code>siteMinCount</code> parameter:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteMinCount <span class="op">=</span> <span class="fl">5</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>This parameter can range from 0 to the number of rows (e.g. taxa)
within your genotype table. Since our example data only has 5 rows, this
prior example will filter out any markers that contain at least 1
missing variant marker site:</p>
<p><img src="../reference/figures/filter_site_05.png"></p>
</div>
<div class="section level3">
<h3 id="using-heterozygosity-frequency">…using heterozygosity frequency<a class="anchor" aria-label="anchor" href="#using-heterozygosity-frequency"></a>
</h3>
<p>Similar to the prior two sections, we can filter by calculated
frequency of heterozygous sites using the parameters,
<code>minHeterozygous</code> and <code>maxHeterozygous</code>. For
example, if we wanted to filter sites containing
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>&gt;</mo><annotation encoding="application/x-tex">&gt;</annotation></semantics></math>
10% heterozygous sites, we could use the following:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        minHeterozygous <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>        maxHeterozygous <span class="op">=</span> <span class="fl">0.1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>which would filter out sites with calculated frequencies:</p>
<p><img src="../reference/figures/filter_site_06.png"></p>
</div>
<div class="section level3">
<h3 id="using-variant-indexes">…using variant indexes<a class="anchor" aria-label="anchor" href="#using-variant-indexes"></a>
</h3>
<p>For the next several sections we will be illustrating how to filter
sites using positional information for each marker. To better explain
these concepts, let’s modify our genotype table (<code>myGT</code>) to
showcase (1) a broader range of positions and (2) different sequence IDs
(i.e. chromosomes). <em>Note</em> Chromosome IDs are not illustrated
this way in TASSEL and are only for demonstration purposes.</p>
<p><img src="../reference/figures/filter_index_01.png"></p>
<p>To filter by site index, we must first set the
<code>siteRangeFilterType</code> parameter to <code>sites</code>. Next,
we will be able to use the parameters, <code>startSite</code> and
<code>endSite</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"sites"</span>,</span>
<span>        startSite <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        endSite <span class="op">=</span> <span class="fl">3</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>In our example above, this will retain indexed sites 1 through 3:</p>
<p><img src="../reference/figures/filter_index_02.png"></p>
</div>
<div class="section level3">
<h3 id="using-physical-positions">…using physical positions<a class="anchor" aria-label="anchor" href="#using-physical-positions"></a>
</h3>
<p>To filter by physical positions, we must first set the
<code>siteRangeFilterType</code> parameter to <code>position</code>.
Next we will be able to use the parameters:</p>
<ul>
<li><code>startChr</code></li>
<li><code>startPos</code></li>
<li><code>endChr</code></li>
<li><code>endPos</code></li>
</ul>
<p>The <code>*Chr</code> parameters will specify which chromosomes or
sequence IDs will remain and <code>*Pos</code> parameters will specify
the physical positions within each chromosome sequence. For example, if
we wanted to filter just chromosome 1, we can specify the following
parameters:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"position"</span>,</span>
<span>        startChr <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        endChr <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        startPos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        endPos <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="../reference/figures/filter_pos_01.png"></p>
<p>If we want to be more specific in our filtering, we can include
positional information as well:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"position"</span>,</span>
<span>        startChr <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        endChr <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        startPos <span class="op">=</span> <span class="fl">250</span>,</span>
<span>        endPos <span class="op">=</span> <span class="fl">700</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="../reference/figures/filter_pos_02.png"></p>
</div>
<div class="section level3">
<h3 id="using-granges-objects">…using <code>GRanges</code> objects<a class="anchor" aria-label="anchor" href="#using-granges-objects"></a>
</h3>
<p>In the prior example, we have manually set parameters to filter by
physical position. The caveat to this process is that you can only
specify a “fixed” range of start and stop positions. For more dynamic
filtering options, we can specify either external files (<em>next
section</em>) or range-based R objects.</p>
<p><code>GRanges</code> class objects from the
<code>GenomicRanges</code> package serve as the standard data structure
for representing genomic locations within the Bioconductor project. A
basic representation of this class entails the following data types:</p>
<ul>
<li>
<code>seqnames</code> - an <code>Rle</code> object of sequence IDs
(e.g. chromsomes)</li>
<li>
<code>ranges</code> - an <code>IRanges</code> object of start and
stop coordinates for range info</li>
<li>
<code>strand</code> - an <code>Rle</code> object representing
sequence strand info (<code>+</code>, <code>-</code>, or
<code>*</code>)</li>
</ul>
<p>For our example, we will represent a <code>GRanges</code> object as
the variable <code>gr</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   250-500      *</span></span>
<span><span class="co">##   [2]     chr2   213-400      *</span></span>
<span><span class="co">##   [3]     chr2   500-700      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>To implement this in our filtration method, we can pass this
<code>GRanges</code> object into our <code>gRangesObj</code> parameters
and setting <code>siteRangeRilterType</code> to <code>"none"</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>        gRangesObj <span class="op">=</span> <span class="va">gr</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>Which will dynamically filter the following marker indexes:</p>
<p><img src="../reference/figures/filter_gr_01.png"></p>
</div>
<div class="section level3">
<h3 id="using-bed-files">…using BED files<a class="anchor" aria-label="anchor" href="#using-bed-files"></a>
</h3>
<p><code>rTASSEL</code> allows filtering for both external <a href="https://m.ensembl.org/info/website/upload/bed.html" class="external-link">BED files</a>
and chromosomal position files. Similar to our <code>GRanges</code>
object, BED files can also represent genomic range information but as a
flat file. In our example, let’s take the same range information and
save it as a BED file:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> cat my_ranges.bed</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="ex">chr1</span>    250   500</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="ex">chr2</span>    213   400</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="ex">chr2</span>    500   700</span></code></pre></div>
<p>This external file can be passed to our filtration method by setting
<code>siteRangeFilterType</code> to <code>"none"</code> and entering the
location of the file on disk in the <code>bedFile</code> parameter:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>        bedFile <span class="op">=</span> <span class="st">"my_ranges.bed"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>This will filter our genotype table as previously shown in
<code>GRanges</code> section:</p>
<p><img src="../reference/figures/filter_gr_01.png"></p>
</div>
<div class="section level3">
<h3 id="using-chromsomal-position-files">…using chromsomal position files<a class="anchor" aria-label="anchor" href="#using-chromsomal-position-files"></a>
</h3>
<p>Similar to a BED file, we can also pass a chromosomal position file.
This is a tab-separated file of two columns:</p>
<ul>
<li>
<code>Chromosome</code> - the sequence or chromosome ID for each
marker</li>
<li>
<code>Position</code> - the physical position (bp) of each
marker</li>
</ul>
<p>In our example, this would look something like this:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> cat my_chr_pos.tsv</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="ex">Chromsome</span>   Position</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="ex">1</span>   300</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="ex">2</span>   213</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="ex">2</span>   665</span></code></pre></div>
<p>This external file can be passed to our filtration method by setting
<code>siteRangeFilterType</code> to <code>"none"</code> and entering the
location of the file on disk in the <code>chrPosFile</code>
parameter:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>        bedFile <span class="op">=</span> <span class="st">"my_chr_pos.tsv"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>This will also filter our genotype table as previously shown in
<code>GRanges</code> section:</p>
<p><img src="../reference/figures/filter_gr_01.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="filtering-by-taxa">Filtering by taxa<a class="anchor" aria-label="anchor" href="#filtering-by-taxa"></a>
</h2>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>In the prior sections we have discussed how to filter
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
observations (i.e.  columns). In the following sections, we will
illustrate ways to filter
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
rows. In order to filter by variant sites (e.g. marker information), we
can use the function <code><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa()</a></code>. This
function was modeled after TASSEL’s <code>FilterTaxaBuilder</code>
plugin to ensure familiarity between the two methods.</p>
<p>While not as complex as site filtration, taxa filtration parameters
can be seperated into the following sections:</p>
<ul>
<li>Genotype information
<ul>
<li><code>minNotMissing</code></li>
<li><code>minHeterozygous</code></li>
<li><code>maxHeterozygous</code></li>
</ul>
</li>
<li>R objects
<ul>
<li><code>taxa</code></li>
</ul>
</li>
</ul>
<p>Since we will be filtering by taxa, let’s represent our toy
illustration to contain actual maize taxa IDs instead of colorful
circles. This is similar to how this would be encoded in TASSEL:</p>
<p><img src="../reference/figures/filter_taxa_00.png"></p>
</div>
<div class="section level3">
<h3 id="using-na-data-1">…using <code>NA</code> data<a class="anchor" aria-label="anchor" href="#using-na-data-1"></a>
</h3>
<p>Similar to site filtration, we can filter by missing sites (encoded
in TASSEL with <code>N</code>). Instead of column (e.g. marker)
filtration, this will filter taxa (rows) based on missing data. Similar
to how we calculated frequency of minor alleles, we can apply the same
logic with missing data with the <code>minNotMissing</code>
parameter:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        minNotMissing <span class="op">=</span> <span class="fl">1.0</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>This parameter is based on the “missingness” frequency calculated for
each row in your genotype table. Since our example data only has 5 rows,
this prior example will filter out any taxa that contain at least 1
missing variant marker site:</p>
<p><img src="../reference/figures/filter_taxa_na.png"></p>
</div>
<div class="section level3">
<h3 id="using-heterozygosity-frequency-1">…using heterozygosity frequency<a class="anchor" aria-label="anchor" href="#using-heterozygosity-frequency-1"></a>
</h3>
<p>We can also filter taxa by calculated frequency of heterozygous sites
using the parameters, <code>minHeterozygous</code> and
<code>maxHeterozygous</code>. For example, if we wanted to filter all
taxa that contain at least one heterozygous site, we could employ the
following parameters:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        minHeterozygous <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>        maxHeterozygous <span class="op">=</span> <span class="fl">0.0</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>which would filter out taxa with calculated frequencies:</p>
<p><img src="../reference/figures/filter_taxa_het.png"></p>
</div>
<div class="section level3">
<h3 id="using-r-objects">…using R objects<a class="anchor" aria-label="anchor" href="#using-r-objects"></a>
</h3>
<p>Filtering select taxa IDs can be achieved by passing an atomic vector
of <code>character</code> data types to the <code>taxa</code> parameter.
In our example, we will select the following IDs:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myFavTaxa</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "B73"  "B97"  "Ky21"</span></span></code></pre>
<p>This vector can passed to the following argument:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        taxa <span class="op">=</span> <span class="va">myFavTaxa</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>Which will filter out any IDs that are <em>not</em> contained in the
previous vector object:</p>
<p><img src="../reference/figures/filter_taxa_vector.png"></p>
<p>This vector logic can applied to various types of subsetting methods
in R. For example, if we had all of the taxa represented in one
vector:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allTaxa</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "B73"   "B97"   "HP301" "Ky21"  "Mo17"</span></span></code></pre>
<p>We could filter from this object using regular expressions. One
common method is using the <code>str_subset</code> method from the
<code>stringr</code> <a href="https://stringr.tidyverse.org/" class="external-link">package</a> in conjunction with
the base R pipe operator <code>|&gt;</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        taxa <span class="op">=</span> <span class="va">allTaxa</span> <span class="op">|&gt;</span> <span class="fu">str_subset</span><span class="op">(</span><span class="st">"^B|^Ky"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>This will create a subsetted vector containing our taxa IDs of
interest in the prior example:</p>
<p><img src="../reference/figures/filter_taxa_vector.png"></p>
<p>While this example might be rather basic, this method is quite
powerful if dealing with a massive breeding panel of several hundred
samples.</p>
</div>
</div>
<div class="section level2">
<h2 id="piping-methods-together">Piping methods together<a class="anchor" aria-label="anchor" href="#piping-methods-together"></a>
</h2>
<div class="section level3">
<h3 id="example-1">Example 1<a class="anchor" aria-label="anchor" href="#example-1"></a>
</h3>
<p>Like with all good practices in R and programming, various
<code>rTASSEL</code> methods can be “piped” together in one continuous
flow of steps. For example, we could filter simultaneously on sites
<em>and</em> taxa using both methods in conjunction with the base R pipe
operator <code>|&gt;</code>. Let’s use our previous toy data set:</p>
<p><img src="../reference/figures/filter_taxa_00.png"></p>
<p>Let’s (1) select taxa IDs that start with the letter “B” or “K”
<em>and</em> (2) select indexed TASSEL marker sites 1 through 3. This
will filter elements shown below:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        taxa <span class="op">=</span> <span class="va">allTaxa</span> <span class="op">|&gt;</span> <span class="fu">str_subset</span><span class="op">(</span><span class="st">"^B|^Ky"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"sites"</span>,</span>
<span>        startSite <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        endSite <span class="op">=</span> <span class="fl">3</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="../reference/figures/filter_taxa_pipe.png"></p>
</div>
<div class="section level3">
<h3 id="example-2">Example 2<a class="anchor" aria-label="anchor" href="#example-2"></a>
</h3>
<p>We can also implement the same methods with other
<code>rTASSEL</code> methods. In this example we will (1) select taxa
IDs that start with the letter “B” or “K”, (2) select indexed TASSEL
marker sites 1 through 3, <em>and</em> (3) export to a VCF flat
file:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myGT</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableTaxa.html">filterGenotypeTableTaxa</a></span><span class="op">(</span></span>
<span>        taxa <span class="op">=</span> <span class="va">allTaxa</span> <span class="op">|&gt;</span> <span class="fu">str_subset</span><span class="op">(</span><span class="st">"^B|^Ky"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span><span class="op">(</span></span>
<span>        siteRangeFilterType <span class="op">=</span> <span class="st">"sites"</span>,</span>
<span>        startSite <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        endSite <span class="op">=</span> <span class="fl">3</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/exportGenotypeTable.html">exportGenotypeTable</a></span><span class="op">(</span></span>
<span>        file <span class="op">=</span> <span class="st">"my_filtered_gt.vcf"</span>,</span>
<span>        format <span class="op">=</span> <span class="st">"vcf"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
