<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>rTASSEL Benchmarks • rTASSEL</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.10/font.css" rel="stylesheet">
<link href="../deps/Roboto_Mono-0.4.10/font.css" rel="stylesheet">
<link href="../deps/Oswald-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="rTASSEL Benchmarks">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rTASSEL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.11.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rTASSEL.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/rtassel_installation.html">Installing rTASSEL</a></li>
    <li><a class="dropdown-item" href="../articles/genotype_filtration.html">Filtering Genotype Tables</a></li>
    <li><a class="dropdown-item" href="../articles/rtassel_benchmarks.html">rTASSEL Benchmarks</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rTASSEL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>rTASSEL Benchmarks</h1>
                        <h4 data-toc-skip class="author">Brandon
Monier</h4>
            
            <h4 data-toc-skip class="date">2026-02-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rTASSEL/blob/fix-coverage-ci/vignettes/rtassel_benchmarks.Rmd" class="external-link"><code>vignettes/rtassel_benchmarks.Rmd</code></a></small>
      <div class="d-none name"><code>rtassel_benchmarks.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>To achieve benchmarking results, several data sets were used.
Genotypic and phenotypic maize data consisting of 279 samples, 3093
variant sites, and 1 measured trait were utilized for the analysis of
variant call format (VCF) import, generalized linear model (GLM)
association, mixed linear model (MLM) association, and kinship
generation times (Flint-Garcia et al., 2005). To illustrate the
effectiveness of the fast association method, 100 simulated RNA
expression traits for the prior genotype data was used. Trait data was
generated using the makeExampleDESeqDataSet() function from the R
package DESeq2 (Love et al., 2014). A larger genotypic data set
consisting of 1,210 samples and 2,255,405 variant sites was also
utilized for large VCF import and kinship generation times. All
benchmarks were generated using the microbenchmark() function from the R
package microbenchmark (Mersmann, 2019).</p>
<p>All benchmarks sans large VCF import and kinship generation times
were evaluated 100 times and recorded on a workstation running 16 GB of
RAM and 4 cores on an Intel® CoreTM i5-6500 CPU with a clock speed of
3.20 GHz and. Large VCF import and kinship generation benchmarks were
evaluated 10 times and recorded on a workstation running 256 GB of RAM
and 12 cores on an Intel® Xeon® CPU E5-2643 v3 with a clock speed of
3.40GHz.</p>
</div>
<div class="section level2">
<h2 id="feature-comparisons">Feature comparisons<a class="anchor" aria-label="anchor" href="#feature-comparisons"></a>
</h2>
<div class="float">
<img src="../reference/figures/benchmarks/pltFeatures.png" alt="Feature comparisons of rTASSEL with other R packages. Features of rTASSEL (y-axis) are compared with other commonly-used R packages (x-axis). Packages that contain a specified feature are highlighted green (yes) and grey (no) if they do not contain a feature or are limited in scope. Association features for packages are based on if said package contains methods for generalized linear models, mixed linear models utilizing the “Q+K” method (Yu et al., 2006), or multi trait fast association methods (Shabalin, 2012). Kinship and distance matrix features denote if a package can return an n x n matrix of values for further use. Packages that contain plotting features indicate if the package contains an automated plot feature instead of using base or grid-based R graphics (R Core Team, 2020) in conjunction with data output. The packages used for this comparison are BGData (Grueneberg and Campos, 2019), GAPIT3 (Wang and Zhang, 2020), MatrixEQTL (Shabalin, 2012), rrBLUP (Endelman, 2011), SNPRelate (Zheng et al., 2012), statgenGWAS (Rossum and Kruijer, 2020), VariantAnnotation (Obenchain et al., 2014), VariantExperiment (Liu et al., 2020), and vcfR (Knaus and Grünwald, 2017)."><div class="figcaption">
<strong>Feature comparisons of rTASSEL with
other R packages.</strong> Features of rTASSEL (y-axis) are compared
with other commonly-used R packages (x-axis). Packages that contain a
specified feature are highlighted green (yes) and grey (no) if they do
not contain a feature or are limited in scope. Association features for
packages are based on if said package contains methods for generalized
linear models, mixed linear models utilizing the “Q+K” method (Yu et
al., 2006), or multi trait fast association methods (Shabalin, 2012).
Kinship and distance matrix features denote if a package can return an n
x n matrix of values for further use. Packages that contain plotting
features indicate if the package contains an automated plot feature
instead of using base or grid-based R graphics (R Core Team, 2020) in
conjunction with data output. The packages used for this comparison are
BGData (Grueneberg and Campos, 2019), GAPIT3 (Wang and Zhang, 2020),
MatrixEQTL (Shabalin, 2012), rrBLUP (Endelman, 2011), SNPRelate (Zheng
et al., 2012), statgenGWAS (Rossum and Kruijer, 2020), VariantAnnotation
(Obenchain et al., 2014), VariantExperiment (Liu et al., 2020), and vcfR
(Knaus and Grünwald, 2017).</div>
</div>
</div>
<div class="section level2">
<h2 id="speed-comparisons">Speed comparisons<a class="anchor" aria-label="anchor" href="#speed-comparisons"></a>
</h2>
<div class="section level3">
<h3 id="vcf-import">VCF import<a class="anchor" aria-label="anchor" href="#vcf-import"></a>
</h3>
<div class="float">
<img src="../reference/figures/benchmarks/pltVCFImportAll.png" alt="VCF import time comparisons of genotypic data. A distribution of replicated benchmark evaluations with recorded means (cross shapes) are plotted for rTASSEL and several R packages: SNPRelate (Zheng et al., 2012), VariantAnnotation (Obenchain et al., 2014), VariantExperiment (Liu et al., 2020), and vcfR (Knaus and Grünwald, 2017). Import times are recorded for 279 samples x 3093 variant sites (A) and 1,210 samples x 2,255,405 variant sites (B)."><div class="figcaption">
<strong>VCF import time comparisons of genotypic
data.</strong> A distribution of replicated benchmark evaluations with
recorded means (cross shapes) are plotted for rTASSEL and several R
packages: SNPRelate (Zheng et al., 2012), VariantAnnotation (Obenchain
et al., 2014), VariantExperiment (Liu et al., 2020), and vcfR (Knaus and
Grünwald, 2017). Import times are recorded for 279 samples x 3093
variant sites (A) and 1,210 samples x 2,255,405 variant sites (B).</div>
</div>
</div>
<div class="section level3">
<h3 id="association-methods">Association methods<a class="anchor" aria-label="anchor" href="#association-methods"></a>
</h3>
<div class="float">
<img src="../reference/figures/benchmarks/pltGLMAssoc.png" alt="GLM association time comparisons. A distribution of replicated benchmark evaluations with recorded means (cross shapes) are plotted for rTASSEL and the R packages GAPIT3 (Wang and Zhang, 2020) and BGData (Grueneberg and Campos, 2019). Import times are recorded for 279 samples x 3093 variant sites and 1 measured trait."><div class="figcaption">
<strong>GLM association time
comparisons.</strong> A distribution of replicated benchmark evaluations
with recorded means (cross shapes) are plotted for rTASSEL and the R
packages GAPIT3 (Wang and Zhang, 2020) and BGData (Grueneberg and
Campos, 2019). Import times are recorded for 279 samples x 3093 variant
sites and 1 measured trait.</div>
</div>
<div class="float">
<img src="../reference/figures/benchmarks/pltMLMAssoc.png" alt="MLM association time comparisons. A distribution of replicated benchmark evaluations with recorded means (cross shapes) are plotted for rTASSEL and the R packages rrBLUP (Endelman, 2011) and GAPIT3 (Wang and Zhang, 2020). Import times are recorded for 279 samples x 3093 variant sites and 1 measured trait."><div class="figcaption">
<strong>MLM association time
comparisons.</strong> A distribution of replicated benchmark evaluations
with recorded means (cross shapes) are plotted for rTASSEL and the R
packages rrBLUP (Endelman, 2011) and GAPIT3 (Wang and Zhang, 2020).
Import times are recorded for 279 samples x 3093 variant sites and 1
measured trait.</div>
</div>
<div class="float">
<img src="../reference/figures/benchmarks/pltFastAssoc.png" alt="Fast association time comparisons. A distribution of replicated benchmark evaluations with recorded means (cross shapes) are plotted for rTASSEL and the R package MatrixEQTL (Shabalin, 2012). Import times are recorded for 279 samples x 3093 variant sites and 100 simulated RNA expression traits."><div class="figcaption">
<strong>Fast association time
comparisons.</strong> A distribution of replicated benchmark evaluations
with recorded means (cross shapes) are plotted for rTASSEL and the R
package MatrixEQTL (Shabalin, 2012). Import times are recorded for 279
samples x 3093 variant sites and 100 simulated RNA expression
traits.</div>
</div>
</div>
<div class="section level3">
<h3 id="kinship-creation">Kinship creation<a class="anchor" aria-label="anchor" href="#kinship-creation"></a>
</h3>
<div class="float">
<img src="../reference/figures/benchmarks/pltKinAll.png" alt="Kinship matrix (IBS) generation time comparisons of genotypic data. A distribution of replicated benchmark evaluations with recorded means (cross shapes) are plotted for rTASSEL and the R packages statgenGWAS (Rossum and Kruijer, 2020) and SNPRelate (Zheng et al., 2012). Generation times are recorded for 279 samples x 3093 variant sites (A) and 1,210 samples x 2,255,405 variant sites (B)."><div class="figcaption">
<strong>Kinship matrix (IBS) generation time
comparisons of genotypic data.</strong> A distribution of replicated
benchmark evaluations with recorded means (cross shapes) are plotted for
rTASSEL and the R packages statgenGWAS (Rossum and Kruijer, 2020) and
SNPRelate (Zheng et al., 2012). Generation times are recorded for 279
samples x 3093 variant sites (A) and 1,210 samples x 2,255,405 variant
sites (B).</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p><b>rTASSEL</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
